// @temperlang/std/json.js
const {
  imul: imul__498
} = globalThis.Math;
import {
  type as type__79, requireInstanceOf as requireInstanceOf__293, clampInt64 as clampInt64__680, cmpFloat as cmpFloat__692, mappedGetOr as mappedGetOr_111, listedGet as listedGet_113, mappedForEach as mappedForEach_125, int64ToInt32 as int64ToInt32_165, int64ToFloat64 as int64ToFloat64_168, float64ToString as float64ToString_175, float64ToInt32 as float64ToInt32_177, float64ToInt64 as float64ToInt64_179, stringToInt32 as stringToInt32_191, stringToFloat64 as stringToFloat64_192, stringToInt64 as stringToInt64_197, listBuilderAdd as listBuilderAdd_209, listedGetOr as listedGetOr_213, listBuilderSet as listBuilderSet_221, listBuilderRemoveLast as listBuilderRemoveLast_226, mapBuilderConstructor as mapBuilderConstructor_287, panic as panic_294, mappedGet as mappedGet_299, mapBuilderSet as mapBuilderSet_300, listBuilderToList as listBuilderToList_306, mappedToMap as mappedToMap_307, stringGet as stringGet_356, stringNext as stringNext_464, stringHasAtLeast as stringHasAtLeast_535, stringBuilderAppendCodePoint as stringBuilderAppendCodePoint_540, requireStringIndex as requireStringIndex_569, int64ToInt32Unsafe as int64ToInt32Unsafe_689
} from "@temperlang/core";
export class InterchangeContext extends type__79() {
  /**
   * @param {string} headerName_78
   * @returns {string | null}
   */
  getHeader(headerName_78) {
    null;
  }
};
export class NullInterchangeContext extends type__79(InterchangeContext) {
  /**
   * @param {string} headerName_81
   * @returns {string | null}
   */
  getHeader(headerName_81) {
    return null;
  }
  /** @type {NullInterchangeContext} */
  static #instance_82 = new NullInterchangeContext();
  /** @returns {NullInterchangeContext} */
  static get instance() {
    return this.#instance_82;
  }
  constructor() {
    super ();
    return;
  }
};
export class JsonProducer extends type__79() {
  startObject() {
    null;
  }
  endObject() {
    null;
  }
  /** @param {string} key_87 */
  objectKey(key_87) {
    null;
  }
  startArray() {
    null;
  }
  endArray() {
    null;
  }
  nullValue() {
    null;
  }
  /** @param {boolean} x_92 */
  booleanValue(x_92) {
    null;
  }
  /** @param {number} x_94 */
  int32Value(x_94) {
    null;
  }
  /** @param {bigint} x_96 */
  int64Value(x_96) {
    null;
  }
  /** @param {number} x_98 */
  float64Value(x_98) {
    null;
  }
  /** @param {string} x_100 */
  numericTokenValue(x_100) {
    null;
  }
  /** @param {string} x_102 */
  stringValue(x_102) {
    null;
  }
  /** @returns {JsonParseErrorReceiver | null} */
  get parseErrorReceiver() {
    return null;
  }
};
export class JsonSyntaxTree extends type__79() {
  /** @param {JsonProducer} p_105 */
  produce(p_105) {
    null;
  }
};
export class JsonObject extends type__79(JsonSyntaxTree) {
  /** @type {Map<string, Array<JsonSyntaxTree>>} */
  #properties_106;
  /**
   * @param {string} propertyKey_108
   * @returns {JsonSyntaxTree | null}
   */
  propertyValueOrNull(propertyKey_108) {
    let return_109;
    const treeList_110 = mappedGetOr_111(this.#properties_106, propertyKey_108, Object.freeze([]));
    const lastIndex_112 = treeList_110.length - 1 | 0;
    if (lastIndex_112 >= 0) {
      return_109 = listedGet_113(treeList_110, lastIndex_112);
    } else {
      return_109 = null;
    }
    return return_109;
  }
  /**
   * @param {string} propertyKey_115
   * @returns {JsonSyntaxTree}
   */
  propertyValueOrBubble(propertyKey_115) {
    let return_116;
    let t_117 = this.propertyValueOrNull(propertyKey_115);
    if (t_117 == null) {
      throw Error();
    } else {
      return_116 = t_117;
    }
    return return_116;
  }
  /** @param {JsonProducer} p_119 */
  produce(p_119) {
    p_119.startObject();
    function fn_120(k_121, vs_122) {
      function fn_123(v_124) {
        p_119.objectKey(k_121);
        v_124.produce(p_119);
        return;
      }
      vs_122.forEach(fn_123);
      return;
    }
    mappedForEach_125(this.#properties_106, fn_120);
    p_119.endObject();
    return;
  }
  /** @param {Map<string, Array<JsonSyntaxTree>>} properties_126 */
  constructor(properties_126) {
    super ();
    this.#properties_106 = properties_126;
    return;
  }
  /** @returns {Map<string, Array<JsonSyntaxTree>>} */
  get properties() {
    return this.#properties_106;
  }
};
export class JsonArray extends type__79(JsonSyntaxTree) {
  /** @type {Array<JsonSyntaxTree>} */
  #elements_128;
  /** @param {JsonProducer} p_130 */
  produce(p_130) {
    p_130.startArray();
    function fn_131(v_132) {
      v_132.produce(p_130);
      return;
    }
    this.#elements_128.forEach(fn_131);
    p_130.endArray();
    return;
  }
  /** @param {Array<JsonSyntaxTree>} elements_133 */
  constructor(elements_133) {
    super ();
    this.#elements_128 = elements_133;
    return;
  }
  /** @returns {Array<JsonSyntaxTree>} */
  get elements() {
    return this.#elements_128;
  }
};
export class JsonBoolean extends type__79(JsonSyntaxTree) {
  /** @type {boolean} */
  #content_135;
  /** @param {JsonProducer} p_137 */
  produce(p_137) {
    p_137.booleanValue(this.#content_135);
    return;
  }
  /** @param {boolean} content_138 */
  constructor(content_138) {
    super ();
    this.#content_135 = content_138;
    return;
  }
  /** @returns {boolean} */
  get content() {
    return this.#content_135;
  }
};
export class JsonNull extends type__79(JsonSyntaxTree) {
  /** @param {JsonProducer} p_141 */
  produce(p_141) {
    p_141.nullValue();
    return;
  }
  constructor() {
    super ();
    return;
  }
};
export class JsonString extends type__79(JsonSyntaxTree) {
  /** @type {string} */
  #content_142;
  /** @param {JsonProducer} p_144 */
  produce(p_144) {
    p_144.stringValue(this.#content_142);
    return;
  }
  /** @param {string} content_145 */
  constructor(content_145) {
    super ();
    this.#content_142 = content_145;
    return;
  }
  /** @returns {string} */
  get content() {
    return this.#content_142;
  }
};
export class JsonNumeric extends type__79(JsonSyntaxTree) {
  /** @returns {string} */
  asJsonNumericToken() {
    null;
  }
  /** @returns {number} */
  asInt32() {
    null;
  }
  /** @returns {bigint} */
  asInt64() {
    null;
  }
  /** @returns {number} */
  asFloat64() {
    null;
  }
};
export class JsonInt32 extends type__79(JsonNumeric) {
  /** @type {number} */
  #content_151;
  /** @param {JsonProducer} p_153 */
  produce(p_153) {
    p_153.int32Value(this.#content_151);
    return;
  }
  /** @returns {string} */
  asJsonNumericToken() {
    return this.#content_151.toString();
  }
  /** @returns {number} */
  asInt32() {
    return this.#content_151;
  }
  /** @returns {bigint} */
  asInt64() {
    return BigInt(this.#content_151);
  }
  /** @returns {number} */
  asFloat64() {
    return this.#content_151;
  }
  /** @param {number} content_158 */
  constructor(content_158) {
    super ();
    this.#content_151 = content_158;
    return;
  }
  /** @returns {number} */
  get content() {
    return this.#content_151;
  }
};
export class JsonInt64 extends type__79(JsonNumeric) {
  /** @type {bigint} */
  #content_160;
  /** @param {JsonProducer} p_162 */
  produce(p_162) {
    p_162.int64Value(this.#content_160);
    return;
  }
  /** @returns {string} */
  asJsonNumericToken() {
    return this.#content_160.toString();
  }
  /** @returns {number} */
  asInt32() {
    return int64ToInt32_165(this.#content_160);
  }
  /** @returns {bigint} */
  asInt64() {
    return this.#content_160;
  }
  /** @returns {number} */
  asFloat64() {
    return int64ToFloat64_168(this.#content_160);
  }
  /** @param {bigint} content_169 */
  constructor(content_169) {
    super ();
    this.#content_160 = content_169;
    return;
  }
  /** @returns {bigint} */
  get content() {
    return this.#content_160;
  }
};
export class JsonFloat64 extends type__79(JsonNumeric) {
  /** @type {number} */
  #content_171;
  /** @param {JsonProducer} p_173 */
  produce(p_173) {
    p_173.float64Value(this.#content_171);
    return;
  }
  /** @returns {string} */
  asJsonNumericToken() {
    return float64ToString_175(this.#content_171);
  }
  /** @returns {number} */
  asInt32() {
    return float64ToInt32_177(this.#content_171);
  }
  /** @returns {bigint} */
  asInt64() {
    return float64ToInt64_179(this.#content_171);
  }
  /** @returns {number} */
  asFloat64() {
    return this.#content_171;
  }
  /** @param {number} content_181 */
  constructor(content_181) {
    super ();
    this.#content_171 = content_181;
    return;
  }
  /** @returns {number} */
  get content() {
    return this.#content_171;
  }
};
export class JsonNumericToken extends type__79(JsonNumeric) {
  /** @type {string} */
  #content_183;
  /** @param {JsonProducer} p_185 */
  produce(p_185) {
    p_185.numericTokenValue(this.#content_183);
    return;
  }
  /** @returns {string} */
  asJsonNumericToken() {
    return this.#content_183;
  }
  /** @returns {number} */
  asInt32() {
    let return_188;
    let t_189;
    let t_190;
    try {
      t_189 = stringToInt32_191(this.#content_183);
      return_188 = t_189;
    } catch {
      t_190 = stringToFloat64_192(this.#content_183);
      return_188 = float64ToInt32_177(t_190);
    }
    return return_188;
  }
  /** @returns {bigint} */
  asInt64() {
    let return_194;
    let t_195;
    let t_196;
    try {
      t_195 = stringToInt64_197(this.#content_183);
      return_194 = t_195;
    } catch {
      t_196 = stringToFloat64_192(this.#content_183);
      return_194 = float64ToInt64_179(t_196);
    }
    return return_194;
  }
  /** @returns {number} */
  asFloat64() {
    return stringToFloat64_192(this.#content_183);
  }
  /** @param {string} content_199 */
  constructor(content_199) {
    super ();
    this.#content_183 = content_199;
    return;
  }
  /** @returns {string} */
  get content() {
    return this.#content_183;
  }
};
export class JsonTextProducer extends type__79(JsonProducer) {
  /** @type {InterchangeContext} */
  #interchangeContext_201;
  /** @type {globalThis.Array<string>} */
  #buffer_202;
  /** @type {Array<number>} */
  #stack_203;
  /** @type {boolean} */
  #wellFormed_204;
  /** @param {InterchangeContext | null} [interchangeContext_205] */
  constructor(interchangeContext_205) {
    super ();
    let interchangeContext_206;
    if (interchangeContext_205 == null) {
      interchangeContext_206 = NullInterchangeContext.instance;
    } else {
      interchangeContext_206 = interchangeContext_205;
    }
    this.#interchangeContext_201 = interchangeContext_206;
    let t_207 = [""];
    this.#buffer_202 = t_207;
    let t_208 = [];
    this.#stack_203 = t_208;
    listBuilderAdd_209(this.#stack_203, 5);
    this.#wellFormed_204 = true;
    return;
  }
  /** @returns {number} */
  #state_211() {
    let t_212 = this.#stack_203.length;
    return listedGetOr_213(this.#stack_203, t_212 - 1 | 0, -1);
  }
  #beforeValue_215() {
    let t_216;
    let t_217;
    let t_218;
    let t_219;
    const currentState_220 = this.#state_211();
    if (currentState_220 === 3) {
      t_216 = this.#stack_203.length;
      listBuilderSet_221(this.#stack_203, t_216 - 1 | 0, 4);
    } else if (currentState_220 === 4) {
      this.#buffer_202[0] += ",";
    } else if (currentState_220 === 1) {
      t_217 = this.#stack_203.length;
      listBuilderSet_221(this.#stack_203, t_217 - 1 | 0, 2);
    } else if (currentState_220 === 5) {
      t_218 = this.#stack_203.length;
      listBuilderSet_221(this.#stack_203, t_218 - 1 | 0, 6);
    } else {
      if (currentState_220 === 6) {
        t_219 = true;
      } else {
        t_219 = currentState_220 === 2;
      }
      if (t_219) {
        this.#wellFormed_204 = false;
      }
    }
    return;
  }
  startObject() {
    this.#beforeValue_215();
    this.#buffer_202[0] += "{";
    listBuilderAdd_209(this.#stack_203, 0);
    return;
  }
  endObject() {
    let t_224;
    this.#buffer_202[0] += "}";
    const currentState_225 = this.#state_211();
    if (0 === currentState_225) {
      t_224 = true;
    } else {
      t_224 = 2 === currentState_225;
    }
    if (t_224) {
      listBuilderRemoveLast_226(this.#stack_203);
    } else {
      this.#wellFormed_204 = false;
    }
    return;
  }
  /** @param {string} key_228 */
  objectKey(key_228) {
    let t_229;
    const currentState_230 = this.#state_211();
    if (!(currentState_230 === 0)) {
      if (currentState_230 === 2) {
        this.#buffer_202[0] += ",";
      } else {
        this.#wellFormed_204 = false;
      }
    }
    encodeJsonString_231(key_228, this.#buffer_202);
    this.#buffer_202[0] += ":";
    if (currentState_230 >= 0) {
      t_229 = this.#stack_203.length;
      listBuilderSet_221(this.#stack_203, t_229 - 1 | 0, 1);
    }
    return;
  }
  startArray() {
    this.#beforeValue_215();
    this.#buffer_202[0] += "[";
    listBuilderAdd_209(this.#stack_203, 3);
    return;
  }
  endArray() {
    let t_234;
    this.#buffer_202[0] += "]";
    const currentState_235 = this.#state_211();
    if (3 === currentState_235) {
      t_234 = true;
    } else {
      t_234 = 4 === currentState_235;
    }
    if (t_234) {
      listBuilderRemoveLast_226(this.#stack_203);
    } else {
      this.#wellFormed_204 = false;
    }
    return;
  }
  nullValue() {
    this.#beforeValue_215();
    this.#buffer_202[0] += "null";
    return;
  }
  /** @param {boolean} x_238 */
  booleanValue(x_238) {
    let t_239;
    this.#beforeValue_215();
    if (x_238) {
      t_239 = "true";
    } else {
      t_239 = "false";
    }
    this.#buffer_202[0] += t_239;
    return;
  }
  /** @param {number} x_241 */
  int32Value(x_241) {
    this.#beforeValue_215();
    let t_242 = x_241.toString();
    this.#buffer_202[0] += t_242;
    return;
  }
  /** @param {bigint} x_244 */
  int64Value(x_244) {
    this.#beforeValue_215();
    let t_245 = x_244.toString();
    this.#buffer_202[0] += t_245;
    return;
  }
  /** @param {number} x_247 */
  float64Value(x_247) {
    this.#beforeValue_215();
    let t_248 = float64ToString_175(x_247);
    this.#buffer_202[0] += t_248;
    return;
  }
  /** @param {string} x_250 */
  numericTokenValue(x_250) {
    this.#beforeValue_215();
    this.#buffer_202[0] += x_250;
    return;
  }
  /** @param {string} x_252 */
  stringValue(x_252) {
    this.#beforeValue_215();
    encodeJsonString_231(x_252, this.#buffer_202);
    return;
  }
  /** @returns {string} */
  toJsonString() {
    let return_254;
    let t_255;
    let t_256;
    let t_257;
    if (this.#wellFormed_204) {
      if (this.#stack_203.length === 1) {
        t_255 = this.#state_211();
        t_256 = t_255 === 6;
      } else {
        t_256 = false;
      }
      t_257 = t_256;
    } else {
      t_257 = false;
    }
    if (t_257) {
      return_254 = this.#buffer_202[0];
    } else {
      throw Error();
    }
    return return_254;
  }
  /** @returns {InterchangeContext} */
  get interchangeContext() {
    return this.#interchangeContext_201;
  }
};
export class JsonParseErrorReceiver extends type__79() {
  /** @param {string} explanation_260 */
  explainJsonError(explanation_260) {
    null;
  }
};
export class JsonSyntaxTreeProducer extends type__79(JsonProducer, JsonParseErrorReceiver) {
  /** @type {Array<Array<JsonSyntaxTree>>} */
  #stack_261;
  /** @type {string | null} */
  #error_262;
  /** @returns {InterchangeContext} */
  get interchangeContext() {
    return NullInterchangeContext.instance;
  }
  constructor() {
    super ();
    let t_264 = [];
    this.#stack_261 = t_264;
    let t_265 = [];
    listBuilderAdd_209(this.#stack_261, t_265);
    this.#error_262 = null;
    return;
  }
  /** @param {JsonSyntaxTree} v_268 */
  #storeValue_267(v_268) {
    let t_269;
    if (! ! this.#stack_261.length) {
      t_269 = this.#stack_261.length;
      listBuilderAdd_209(listedGet_113(this.#stack_261, t_269 - 1 | 0), v_268);
    }
    return;
  }
  startObject() {
    let t_271 = [];
    listBuilderAdd_209(this.#stack_261, t_271);
    return;
  }
  endObject() {
    let return_273;
    let t_274;
    let t_275;
    let t_276;
    let t_277;
    let t_278;
    let t_279;
    let t_280;
    let t_281;
    let t_282;
    let t_283;
    fn_284: {
      if (! this.#stack_261.length) {
        return_273 = void 0;
        break fn_284;
      }
      const ls_285 = listBuilderRemoveLast_226(this.#stack_261);
      const m_286 = mapBuilderConstructor_287();
      let multis_288 = null;
      let i_289 = 0;
      let n_290 = ls_285.length & -2;
      while (i_289 < n_290) {
        const postfixReturn_291 = i_289;
        i_289 = i_289 + 1 | 0;
        const keyTree_292 = listedGet_113(ls_285, postfixReturn_291);
        if (!(keyTree_292 instanceof JsonString)) {
          break;
        }
        try {
          t_276 = requireInstanceOf__293(keyTree_292, JsonString);
          t_277 = t_276;
        } catch {
          t_277 = panic_294();
        }
        const key_295 = t_277.content;
        const postfixReturn_296 = i_289;
        i_289 = i_289 + 1 | 0;
        const value_297 = listedGet_113(ls_285, postfixReturn_296);
        if (m_286.has(key_295)) {
          if (multis_288 == null) {
            t_274 = mapBuilderConstructor_287();
            multis_288 = t_274;
          }
          try {
            if (multis_288 == null) {
              throw Error();
            } else {
              t_278 = multis_288;
            }
            t_279 = t_278;
          } catch {
            t_279 = panic_294();
          }
          const mb_298 = t_279;
          if (! mb_298.has(key_295)) {
            try {
              t_280 = mappedGet_299(m_286, key_295);
              t_281 = t_280;
            } catch {
              t_281 = panic_294();
            }
            mapBuilderSet_300(mb_298, key_295, t_281.slice());
          }
          try {
            t_282 = mappedGet_299(mb_298, key_295);
            t_283 = t_282;
          } catch {
            t_283 = panic_294();
          }
          listBuilderAdd_209(t_283, value_297);
        } else {
          mapBuilderSet_300(m_286, key_295, Object.freeze([value_297]));
        }
      }
      const multis_301 = multis_288;
      if (!(multis_301 == null)) {
        function fn_302(k_303, vs_304) {
          let t_305 = listBuilderToList_306(vs_304);
          mapBuilderSet_300(m_286, k_303, t_305);
          return;
        }
        mappedForEach_125(multis_301, fn_302);
      }
      t_275 = new JsonObject(mappedToMap_307(m_286));
      this.#storeValue_267(t_275);
      return_273 = void 0;
    }
    return return_273;
  }
  /** @param {string} key_309 */
  objectKey(key_309) {
    let t_310 = new JsonString(key_309);
    this.#storeValue_267(t_310);
    return;
  }
  startArray() {
    let t_312 = [];
    listBuilderAdd_209(this.#stack_261, t_312);
    return;
  }
  endArray() {
    let return_314;
    let t_315;
    fn_316: {
      if (! this.#stack_261.length) {
        return_314 = void 0;
        break fn_316;
      }
      const ls_317 = listBuilderRemoveLast_226(this.#stack_261);
      t_315 = new JsonArray(listBuilderToList_306(ls_317));
      this.#storeValue_267(t_315);
      return_314 = void 0;
    }
    return return_314;
  }
  nullValue() {
    let t_319 = new JsonNull();
    this.#storeValue_267(t_319);
    return;
  }
  /** @param {boolean} x_321 */
  booleanValue(x_321) {
    let t_322 = new JsonBoolean(x_321);
    this.#storeValue_267(t_322);
    return;
  }
  /** @param {number} x_324 */
  int32Value(x_324) {
    let t_325 = new JsonInt32(x_324);
    this.#storeValue_267(t_325);
    return;
  }
  /** @param {bigint} x_327 */
  int64Value(x_327) {
    let t_328 = new JsonInt64(x_327);
    this.#storeValue_267(t_328);
    return;
  }
  /** @param {number} x_330 */
  float64Value(x_330) {
    let t_331 = new JsonFloat64(x_330);
    this.#storeValue_267(t_331);
    return;
  }
  /** @param {string} x_333 */
  numericTokenValue(x_333) {
    let t_334 = new JsonNumericToken(x_333);
    this.#storeValue_267(t_334);
    return;
  }
  /** @param {string} x_336 */
  stringValue(x_336) {
    let t_337 = new JsonString(x_336);
    this.#storeValue_267(t_337);
    return;
  }
  /** @returns {JsonSyntaxTree} */
  toJsonSyntaxTree() {
    let t_339;
    if (this.#stack_261.length !== 1) {
      t_339 = true;
    } else {
      t_339 = !(this.#error_262 == null);
    }
    if (t_339) {
      throw Error();
    }
    const ls_340 = listedGet_113(this.#stack_261, 0);
    if (ls_340.length !== 1) {
      throw Error();
    }
    return listedGet_113(ls_340, 0);
  }
  /** @returns {string | null} */
  get jsonError() {
    return this.#error_262;
  }
  /** @returns {JsonParseErrorReceiver} */
  get parseErrorReceiver() {
    return this;
  }
  /** @param {string} error_344 */
  explainJsonError(error_344) {
    this.#error_262 = error_344;
    return;
  }
};
/**
 * @param {string} sourceText_346
 * @param {globalThis.number} i_347
 * @param {JsonProducer} out_348
 * @returns {globalThis.number}
 */
function parseJsonValue_345(sourceText_346, i_347, out_348) {
  let return_349;
  let t_350;
  let t_351;
  let t_352;
  fn_353: {
    t_350 = skipJsonSpaces_354(sourceText_346, i_347);
    i_347 = t_350;
    if (!(sourceText_346.length > i_347)) {
      expectedTokenError_355(sourceText_346, i_347, out_348, "JSON value");
      return_349 = -1;
      break fn_353;
    }
    t_351 = stringGet_356(sourceText_346, i_347);
    if (t_351 === 123) {
      return_349 = parseJsonObject_357(sourceText_346, i_347, out_348);
    } else if (t_351 === 91) {
      return_349 = parseJsonArray_358(sourceText_346, i_347, out_348);
    } else if (t_351 === 34) {
      return_349 = parseJsonString_359(sourceText_346, i_347, out_348);
    } else {
      if (t_351 === 116) {
        t_352 = true;
      } else {
        t_352 = t_351 === 102;
      }
      if (t_352) {
        return_349 = parseJsonBoolean_360(sourceText_346, i_347, out_348);
      } else if (t_351 === 110) {
        return_349 = parseJsonNull_361(sourceText_346, i_347, out_348);
      } else {
        return_349 = parseJsonNumber_362(sourceText_346, i_347, out_348);
      }
    }
  }
  return return_349;
}
/** @template T_363 */
export class JsonAdapter extends type__79() {
  /**
   * @param {T_363} x_365
   * @param {JsonProducer} p_366
   */
  encodeToJson(x_365, p_366) {
    null;
  }
  /**
   * @param {JsonSyntaxTree} t_368
   * @param {InterchangeContext} ic_369
   * @returns {T_363}
   */
  decodeFromJson(t_368, ic_369) {
    null;
  }
};
class BooleanJsonAdapter_370 extends type__79(JsonAdapter) {
  /**
   * @param {boolean} x_372
   * @param {JsonProducer} p_373
   */
  encodeToJson(x_372, p_373) {
    p_373.booleanValue(x_372);
    return;
  }
  /**
   * @param {JsonSyntaxTree} t_375
   * @param {InterchangeContext} ic_376
   * @returns {boolean}
   */
  decodeFromJson(t_375, ic_376) {
    let t_377;
    t_377 = requireInstanceOf__293(t_375, JsonBoolean);
    return t_377.content;
  }
  constructor() {
    super ();
    return;
  }
}
class Float64JsonAdapter_378 extends type__79(JsonAdapter) {
  /**
   * @param {number} x_380
   * @param {JsonProducer} p_381
   */
  encodeToJson(x_380, p_381) {
    p_381.float64Value(x_380);
    return;
  }
  /**
   * @param {JsonSyntaxTree} t_383
   * @param {InterchangeContext} ic_384
   * @returns {number}
   */
  decodeFromJson(t_383, ic_384) {
    let t_385;
    t_385 = requireInstanceOf__293(t_383, JsonNumeric);
    return t_385.asFloat64();
  }
  constructor() {
    super ();
    return;
  }
}
class Int32JsonAdapter_386 extends type__79(JsonAdapter) {
  /**
   * @param {number} x_388
   * @param {JsonProducer} p_389
   */
  encodeToJson(x_388, p_389) {
    p_389.int32Value(x_388);
    return;
  }
  /**
   * @param {JsonSyntaxTree} t_391
   * @param {InterchangeContext} ic_392
   * @returns {number}
   */
  decodeFromJson(t_391, ic_392) {
    let t_393;
    t_393 = requireInstanceOf__293(t_391, JsonNumeric);
    return t_393.asInt32();
  }
  constructor() {
    super ();
    return;
  }
}
class Int64JsonAdapter_394 extends type__79(JsonAdapter) {
  /**
   * @param {bigint} x_396
   * @param {JsonProducer} p_397
   */
  encodeToJson(x_396, p_397) {
    p_397.int64Value(x_396);
    return;
  }
  /**
   * @param {JsonSyntaxTree} t_399
   * @param {InterchangeContext} ic_400
   * @returns {bigint}
   */
  decodeFromJson(t_399, ic_400) {
    let t_401;
    t_401 = requireInstanceOf__293(t_399, JsonNumeric);
    return t_401.asInt64();
  }
  constructor() {
    super ();
    return;
  }
}
class StringJsonAdapter_402 extends type__79(JsonAdapter) {
  /**
   * @param {string} x_404
   * @param {JsonProducer} p_405
   */
  encodeToJson(x_404, p_405) {
    p_405.stringValue(x_404);
    return;
  }
  /**
   * @param {JsonSyntaxTree} t_407
   * @param {InterchangeContext} ic_408
   * @returns {string}
   */
  decodeFromJson(t_407, ic_408) {
    let t_409;
    t_409 = requireInstanceOf__293(t_407, JsonString);
    return t_409.content;
  }
  constructor() {
    super ();
    return;
  }
}
/** @template T_411 */
class ListJsonAdapter_410 extends type__79(JsonAdapter) {
  /** @type {JsonAdapter<T_411>} */
  #adapterForT_412;
  /**
   * @param {Array<T_411>} x_414
   * @param {JsonProducer} p_415
   */
  encodeToJson(x_414, p_415) {
    const this418 = this;
    p_415.startArray();
    function fn_416(el_417) {
      this418.#adapterForT_412.encodeToJson(el_417, p_415);
      return;
    }
    x_414.forEach(fn_416);
    p_415.endArray();
    return;
  }
  /**
   * @param {JsonSyntaxTree} t_420
   * @param {InterchangeContext} ic_421
   * @returns {Array<T_411>}
   */
  decodeFromJson(t_420, ic_421) {
    let t_422;
    const b_423 = [];
    let t_424;
    t_424 = requireInstanceOf__293(t_420, JsonArray);
    const elements_425 = t_424.elements;
    const n_426 = elements_425.length;
    let i_427 = 0;
    while (i_427 < n_426) {
      const el_428 = listedGet_113(elements_425, i_427);
      i_427 = i_427 + 1 | 0;
      t_422 = this.#adapterForT_412.decodeFromJson(el_428, ic_421);
      listBuilderAdd_209(b_423, t_422);
    }
    return listBuilderToList_306(b_423);
  }
  /** @param {JsonAdapter<T_411>} adapterForT_429 */
  constructor(adapterForT_429) {
    super ();
    this.#adapterForT_412 = adapterForT_429;
    return;
  }
}
/** @template T_430 */
export class OrNullJsonAdapter extends type__79(JsonAdapter) {
  /** @type {JsonAdapter<T_430>} */
  #adapterForT_431;
  /**
   * @param {T_430 | null} x_433
   * @param {JsonProducer} p_434
   */
  encodeToJson(x_433, p_434) {
    if (x_433 == null) {
      p_434.nullValue();
    } else {
      const x_435 = x_433;
      this.#adapterForT_431.encodeToJson(x_435, p_434);
    }
    return;
  }
  /**
   * @param {JsonSyntaxTree} t_437
   * @param {InterchangeContext} ic_438
   * @returns {T_430 | null}
   */
  decodeFromJson(t_437, ic_438) {
    let return_439;
    if (t_437 instanceof JsonNull) {
      return_439 = null;
    } else {
      return_439 = this.#adapterForT_431.decodeFromJson(t_437, ic_438);
    }
    return return_439;
  }
  /** @param {JsonAdapter<T_430>} adapterForT_440 */
  constructor(adapterForT_440) {
    super ();
    this.#adapterForT_431 = adapterForT_440;
    return;
  }
};
/** @type {Array<string>} */
const hexDigits_441 = Object.freeze(["0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "a", "b", "c", "d", "e", "f"]);
/**
 * @param {number} cp_443
 * @param {globalThis.Array<string>} buffer_444
 */
function encodeHex4_442(cp_443, buffer_444) {
  const b0_445 = (cp_443 / 4096 | 0) & 15;
  const b1_446 = (cp_443 / 256 | 0) & 15;
  const b2_447 = (cp_443 / 16 | 0) & 15;
  const b3_448 = cp_443 & 15;
  let t_449 = listedGet_113(hexDigits_441, b0_445);
  buffer_444[0] += t_449;
  let t_450 = listedGet_113(hexDigits_441, b1_446);
  buffer_444[0] += t_450;
  let t_451 = listedGet_113(hexDigits_441, b2_447);
  buffer_444[0] += t_451;
  let t_452 = listedGet_113(hexDigits_441, b3_448);
  buffer_444[0] += t_452;
  return;
}
/**
 * @param {string} x_453
 * @param {globalThis.Array<string>} buffer_454
 */
function encodeJsonString_231(x_453, buffer_454) {
  let t_455;
  let t_456;
  let t_457;
  let t_458;
  buffer_454[0] += "\u0022";
  let i_459 = 0;
  let emitted_460 = i_459;
  while (true) {
    if (!(x_453.length > i_459)) {
      break;
    }
    const cp_461 = stringGet_356(x_453, i_459);
    if (cp_461 === 8) {
      t_458 = "\\b";
    } else if (cp_461 === 9) {
      t_458 = "\\t";
    } else if (cp_461 === 10) {
      t_458 = "\\n";
    } else if (cp_461 === 12) {
      t_458 = "\\f";
    } else if (cp_461 === 13) {
      t_458 = "\\r";
    } else if (cp_461 === 34) {
      t_458 = "\\\u0022";
    } else if (cp_461 === 92) {
      t_458 = "\\\\";
    } else {
      if (cp_461 < 32) {
        t_456 = true;
      } else {
        if (55296 <= cp_461) {
          t_455 = cp_461 <= 57343;
        } else {
          t_455 = false;
        }
        t_456 = t_455;
      }
      if (t_456) {
        t_457 = "\\u";
      } else {
        t_457 = "";
      }
      t_458 = t_457;
    }
    const replacement_462 = t_458;
    const nextI_463 = stringNext_464(x_453, i_459);
    if (replacement_462 !== "") {
      buffer_454[0] += x_453.substring(emitted_460, i_459);
      buffer_454[0] += replacement_462;
      if (replacement_462 === "\\u") {
        encodeHex4_442(cp_461, buffer_454);
      }
      emitted_460 = nextI_463;
    }
    i_459 = nextI_463;
  }
  buffer_454[0] += x_453.substring(emitted_460, i_459);
  buffer_454[0] += "\u0022";
  return;
}
/**
 * @param {JsonProducer} out_466
 * @param {string} explanation_467
 */
function storeJsonError_465(out_466, explanation_467) {
  let t_468 = out_466.parseErrorReceiver;
  if (!(t_468 == null)) {
    t_468.explainJsonError(explanation_467);
  }
  return;
}
/**
 * @param {string} sourceText_469
 * @param {globalThis.number} i_470
 * @param {JsonProducer} out_471
 * @param {string} shortExplanation_472
 */
function expectedTokenError_355(sourceText_469, i_470, out_471, shortExplanation_472) {
  let t_473;
  let t_474;
  let gotten_475;
  if (sourceText_469.length > i_470) {
    t_473 = sourceText_469.length;
    t_474 = sourceText_469.substring(i_470, t_473);
    gotten_475 = "`" + t_474 + "`";
  } else {
    gotten_475 = "end-of-file";
  }
  storeJsonError_465(out_471, "Expected " + shortExplanation_472 + ", but got " + gotten_475);
  return;
}
/**
 * @param {string} sourceText_476
 * @param {globalThis.number} i_477
 * @returns {globalThis.number}
 */
function skipJsonSpaces_354(sourceText_476, i_477) {
  let t_478;
  let t_479;
  let t_480;
  let t_481;
  let t_482;
  while (true) {
    if (!(sourceText_476.length > i_477)) {
      break;
    }
    t_478 = stringGet_356(sourceText_476, i_477);
    if (t_478 === 9) {
      t_482 = true;
    } else {
      if (t_478 === 10) {
        t_481 = true;
      } else {
        if (t_478 === 13) {
          t_480 = true;
        } else {
          t_480 = t_478 === 32;
        }
        t_481 = t_480;
      }
      t_482 = t_481;
    }
    if (! t_482) {
      break;
    }
    t_479 = stringNext_464(sourceText_476, i_477);
    i_477 = t_479;
  }
  return i_477;
}
/**
 * @param {string} sourceText_484
 * @param {globalThis.number} start_485
 * @param {globalThis.number} limit_486
 * @returns {number}
 */
function decodeHexUnsigned_483(sourceText_484, start_485, limit_486) {
  let return_487;
  let t_488;
  let t_489;
  let t_490;
  let t_491;
  let t_492;
  fn_493: {
    let n_494 = 0;
    let i_495 = start_485;
    while (true) {
      if (!(i_495 - limit_486 < 0)) {
        break;
      }
      const cp_496 = stringGet_356(sourceText_484, i_495);
      if (48 <= cp_496) {
        t_489 = cp_496 <= 48;
      } else {
        t_489 = false;
      }
      if (t_489) {
        t_492 = cp_496 - 48 | 0;
      } else {
        if (65 <= cp_496) {
          t_490 = cp_496 <= 70;
        } else {
          t_490 = false;
        }
        if (t_490) {
          t_492 = (cp_496 - 65 | 0) + 10 | 0;
        } else {
          if (97 <= cp_496) {
            t_491 = cp_496 <= 102;
          } else {
            t_491 = false;
          }
          if (t_491) {
            t_492 = (cp_496 - 97 | 0) + 10 | 0;
          } else {
            return_487 = -1;
            break fn_493;
          }
        }
      }
      const digit_497 = t_492;
      n_494 = imul__498(n_494, 16) + digit_497 | 0;
      t_488 = stringNext_464(sourceText_484, i_495);
      i_495 = t_488;
    }
    return_487 = n_494;
  }
  return return_487;
}
/**
 * @param {string} sourceText_500
 * @param {globalThis.number} i_501
 * @param {globalThis.Array<string>} sb_502
 * @param {JsonProducer} errOut_503
 * @returns {globalThis.number}
 */
function parseJsonStringTo_499(sourceText_500, i_501, sb_502, errOut_503) {
  let return_504;
  let t_505;
  let t_506;
  let t_507;
  let t_508;
  let t_509;
  let t_510;
  let t_511;
  let t_512;
  let t_513;
  let t_514;
  let t_515;
  let t_516;
  let t_517;
  let t_518;
  let t_519;
  let t_520;
  let t_521;
  let t_522;
  let t_523;
  let t_524;
  let t_525;
  let t_526;
  fn_527: {
    if (!(sourceText_500.length > i_501)) {
      t_516 = true;
    } else {
      t_505 = stringGet_356(sourceText_500, i_501);
      t_516 = t_505 !== 34;
    }
    if (t_516) {
      expectedTokenError_355(sourceText_500, i_501, errOut_503, "\u0022");
      return_504 = -1;
      break fn_527;
    }
    t_506 = stringNext_464(sourceText_500, i_501);
    i_501 = t_506;
    let leadSurrogate_528 = -1;
    let consumed_529 = i_501;
    while (true) {
      if (!(sourceText_500.length > i_501)) {
        break;
      }
      const cp_530 = stringGet_356(sourceText_500, i_501);
      if (cp_530 === 34) {
        break;
      }
      t_507 = stringNext_464(sourceText_500, i_501);
      let iNext_531 = t_507;
      const end_532 = sourceText_500.length;
      let needToFlush_533 = false;
      if (cp_530 !== 92) {
        t_522 = cp_530;
      } else {
        needToFlush_533 = true;
        if (!(sourceText_500.length > iNext_531)) {
          expectedTokenError_355(sourceText_500, iNext_531, errOut_503, "escape sequence");
          return_504 = -1;
          break fn_527;
        }
        const esc0_534 = stringGet_356(sourceText_500, iNext_531);
        t_508 = stringNext_464(sourceText_500, iNext_531);
        iNext_531 = t_508;
        if (esc0_534 === 34) {
          t_518 = true;
        } else {
          if (esc0_534 === 92) {
            t_517 = true;
          } else {
            t_517 = esc0_534 === 47;
          }
          t_518 = t_517;
        }
        if (t_518) {
          t_521 = esc0_534;
        } else if (esc0_534 === 98) {
          t_521 = 8;
        } else if (esc0_534 === 102) {
          t_521 = 12;
        } else if (esc0_534 === 110) {
          t_521 = 10;
        } else if (esc0_534 === 114) {
          t_521 = 13;
        } else if (esc0_534 === 116) {
          t_521 = 9;
        } else if (esc0_534 === 117) {
          if (stringHasAtLeast_535(sourceText_500, iNext_531, end_532, 4)) {
            const startHex_536 = iNext_531;
            t_509 = stringNext_464(sourceText_500, iNext_531);
            iNext_531 = t_509;
            t_510 = stringNext_464(sourceText_500, iNext_531);
            iNext_531 = t_510;
            t_511 = stringNext_464(sourceText_500, iNext_531);
            iNext_531 = t_511;
            t_512 = stringNext_464(sourceText_500, iNext_531);
            iNext_531 = t_512;
            t_513 = decodeHexUnsigned_483(sourceText_500, startHex_536, iNext_531);
            t_519 = t_513;
          } else {
            t_519 = -1;
          }
          const hex_537 = t_519;
          if (hex_537 < 0) {
            expectedTokenError_355(sourceText_500, iNext_531, errOut_503, "four hex digits");
            return_504 = -1;
            break fn_527;
          }
          t_520 = hex_537;
          t_521 = t_520;
        } else {
          expectedTokenError_355(sourceText_500, iNext_531, errOut_503, "escape sequence");
          return_504 = -1;
          break fn_527;
        }
        t_522 = t_521;
      }
      let decodedCp_538 = t_522;
      if (leadSurrogate_528 >= 0) {
        needToFlush_533 = true;
        const lead_539 = leadSurrogate_528;
        if (56320 <= decodedCp_538) {
          t_523 = decodedCp_538 <= 57343;
        } else {
          t_523 = false;
        }
        if (t_523) {
          leadSurrogate_528 = -1;
          decodedCp_538 = 65536 +(imul__498(lead_539 - 55296 | 0, 1024) |(decodedCp_538 - 56320 | 0)) | 0;
        }
      } else {
        if (55296 <= decodedCp_538) {
          t_524 = decodedCp_538 <= 56319;
        } else {
          t_524 = false;
        }
        if (t_524) {
          needToFlush_533 = true;
        }
      }
      if (needToFlush_533) {
        sb_502[0] += sourceText_500.substring(consumed_529, i_501);
        if (leadSurrogate_528 >= 0) {
          try {
            stringBuilderAppendCodePoint_540(sb_502, leadSurrogate_528);
          } catch {
            throw Error();
          }
        }
        if (55296 <= decodedCp_538) {
          t_525 = decodedCp_538 <= 56319;
        } else {
          t_525 = false;
        }
        if (t_525) {
          leadSurrogate_528 = decodedCp_538;
        } else {
          leadSurrogate_528 = -1;
          try {
            stringBuilderAppendCodePoint_540(sb_502, decodedCp_538);
          } catch {
            throw Error();
          }
        }
        consumed_529 = iNext_531;
      }
      i_501 = iNext_531;
    }
    if (!(sourceText_500.length > i_501)) {
      t_526 = true;
    } else {
      t_514 = stringGet_356(sourceText_500, i_501);
      t_526 = t_514 !== 34;
    }
    if (t_526) {
      expectedTokenError_355(sourceText_500, i_501, errOut_503, "\u0022");
      return_504 = -1;
    } else {
      if (leadSurrogate_528 >= 0) {
        try {
          stringBuilderAppendCodePoint_540(sb_502, leadSurrogate_528);
        } catch {
          throw Error();
        }
      } else {
        sb_502[0] += sourceText_500.substring(consumed_529, i_501);
      }
      t_515 = stringNext_464(sourceText_500, i_501);
      i_501 = t_515;
      return_504 = i_501;
    }
  }
  return return_504;
}
/**
 * @param {string} sourceText_541
 * @param {globalThis.number} i_542
 * @param {JsonProducer} out_543
 * @returns {globalThis.number}
 */
function parseJsonObject_357(sourceText_541, i_542, out_543) {
  let return_544;
  let t_545;
  let t_546;
  let t_547;
  let t_548;
  let t_549;
  let t_550;
  let t_551;
  let t_552;
  let t_553;
  let t_554;
  let t_555;
  let t_556;
  let t_557;
  let t_558;
  let t_559;
  let t_560;
  let t_561;
  let t_562;
  let t_563;
  let t_564;
  let t_565;
  fn_566: {
    try {
      if (!(sourceText_541.length > i_542)) {
        t_558 = true;
      } else {
        t_545 = stringGet_356(sourceText_541, i_542);
        t_558 = t_545 !== 123;
      }
      if (t_558) {
        expectedTokenError_355(sourceText_541, i_542, out_543, "'{'");
        return_544 = -1;
        break fn_566;
      }
      out_543.startObject();
      t_546 = stringNext_464(sourceText_541, i_542);
      t_547 = skipJsonSpaces_354(sourceText_541, t_546);
      i_542 = t_547;
      if (sourceText_541.length > i_542) {
        t_548 = stringGet_356(sourceText_541, i_542);
        t_559 = t_548 !== 125;
      } else {
        t_559 = false;
      }
      if (t_559) {
        while (true) {
          const keyBuffer_567 = [""];
          const afterKey_568 = parseJsonStringTo_499(sourceText_541, i_542, keyBuffer_567, out_543);
          if (!(afterKey_568 >= 0)) {
            return_544 = -1;
            break fn_566;
          }
          t_549 = keyBuffer_567[0];
          out_543.objectKey(t_549);
          try {
            t_560 = requireStringIndex_569(afterKey_568);
            t_561 = t_560;
          } catch {
            t_561 = panic_294();
          }
          t_550 = skipJsonSpaces_354(sourceText_541, t_561);
          i_542 = t_550;
          if (sourceText_541.length > i_542) {
            t_551 = stringGet_356(sourceText_541, i_542);
            t_562 = t_551 === 58;
          } else {
            t_562 = false;
          }
          if (t_562) {
            t_552 = stringNext_464(sourceText_541, i_542);
            i_542 = t_552;
            const afterPropertyValue_570 = parseJsonValue_345(sourceText_541, i_542, out_543);
            if (!(afterPropertyValue_570 >= 0)) {
              return_544 = -1;
              break fn_566;
            }
            t_563 = requireStringIndex_569(afterPropertyValue_570);
            i_542 = t_563;
          } else {
            expectedTokenError_355(sourceText_541, i_542, out_543, "':'");
            return_544 = -1;
            break fn_566;
          }
          t_553 = skipJsonSpaces_354(sourceText_541, i_542);
          i_542 = t_553;
          if (sourceText_541.length > i_542) {
            t_554 = stringGet_356(sourceText_541, i_542);
            t_564 = t_554 === 44;
          } else {
            t_564 = false;
          }
          if (t_564) {
            t_555 = stringNext_464(sourceText_541, i_542);
            t_556 = skipJsonSpaces_354(sourceText_541, t_555);
            i_542 = t_556;
          } else {
            break;
          }
        }
      }
      if (sourceText_541.length > i_542) {
        t_557 = stringGet_356(sourceText_541, i_542);
        t_565 = t_557 === 125;
      } else {
        t_565 = false;
      }
      if (t_565) {
        out_543.endObject();
        return_544 = stringNext_464(sourceText_541, i_542);
      } else {
        expectedTokenError_355(sourceText_541, i_542, out_543, "'}'");
        return_544 = -1;
      }
    } catch {
      return_544 = panic_294();
    }
  }
  return return_544;
}
/**
 * @param {string} sourceText_571
 * @param {globalThis.number} i_572
 * @param {JsonProducer} out_573
 * @returns {globalThis.number}
 */
function parseJsonArray_358(sourceText_571, i_572, out_573) {
  let return_574;
  let t_575;
  let t_576;
  let t_577;
  let t_578;
  let t_579;
  let t_580;
  let t_581;
  let t_582;
  let t_583;
  let t_584;
  let t_585;
  let t_586;
  let t_587;
  let t_588;
  fn_589: {
    try {
      if (!(sourceText_571.length > i_572)) {
        t_584 = true;
      } else {
        t_575 = stringGet_356(sourceText_571, i_572);
        t_584 = t_575 !== 91;
      }
      if (t_584) {
        expectedTokenError_355(sourceText_571, i_572, out_573, "'['");
        return_574 = -1;
        break fn_589;
      }
      out_573.startArray();
      t_576 = stringNext_464(sourceText_571, i_572);
      t_577 = skipJsonSpaces_354(sourceText_571, t_576);
      i_572 = t_577;
      if (sourceText_571.length > i_572) {
        t_578 = stringGet_356(sourceText_571, i_572);
        t_585 = t_578 !== 93;
      } else {
        t_585 = false;
      }
      if (t_585) {
        while (true) {
          const afterElementValue_590 = parseJsonValue_345(sourceText_571, i_572, out_573);
          if (!(afterElementValue_590 >= 0)) {
            return_574 = -1;
            break fn_589;
          }
          t_586 = requireStringIndex_569(afterElementValue_590);
          i_572 = t_586;
          t_579 = skipJsonSpaces_354(sourceText_571, i_572);
          i_572 = t_579;
          if (sourceText_571.length > i_572) {
            t_580 = stringGet_356(sourceText_571, i_572);
            t_587 = t_580 === 44;
          } else {
            t_587 = false;
          }
          if (t_587) {
            t_581 = stringNext_464(sourceText_571, i_572);
            t_582 = skipJsonSpaces_354(sourceText_571, t_581);
            i_572 = t_582;
          } else {
            break;
          }
        }
      }
      if (sourceText_571.length > i_572) {
        t_583 = stringGet_356(sourceText_571, i_572);
        t_588 = t_583 === 93;
      } else {
        t_588 = false;
      }
      if (t_588) {
        out_573.endArray();
        return_574 = stringNext_464(sourceText_571, i_572);
      } else {
        expectedTokenError_355(sourceText_571, i_572, out_573, "']'");
        return_574 = -1;
      }
    } catch {
      return_574 = panic_294();
    }
  }
  return return_574;
}
/**
 * @param {string} sourceText_591
 * @param {globalThis.number} i_592
 * @param {JsonProducer} out_593
 * @returns {globalThis.number}
 */
function parseJsonString_359(sourceText_591, i_592, out_593) {
  let t_594;
  const sb_595 = [""];
  const after_596 = parseJsonStringTo_499(sourceText_591, i_592, sb_595, out_593);
  if (after_596 >= 0) {
    t_594 = sb_595[0];
    out_593.stringValue(t_594);
  }
  return after_596;
}
/**
 * @param {string} string_598
 * @param {globalThis.number} inString_599
 * @param {string} substring_600
 * @returns {globalThis.number}
 */
function afterSubstring_597(string_598, inString_599, substring_600) {
  let return_601;
  let t_602;
  let t_603;
  fn_604: {
    let i_605 = inString_599;
    let j_606 = 0;
    while (true) {
      if (!(substring_600.length > j_606)) {
        break;
      }
      if (!(string_598.length > i_605)) {
        return_601 = -1;
        break fn_604;
      }
      if (stringGet_356(string_598, i_605) !== stringGet_356(substring_600, j_606)) {
        return_601 = -1;
        break fn_604;
      }
      t_602 = stringNext_464(string_598, i_605);
      i_605 = t_602;
      t_603 = stringNext_464(substring_600, j_606);
      j_606 = t_603;
    }
    return_601 = i_605;
  }
  return return_601;
}
/**
 * @param {string} sourceText_607
 * @param {globalThis.number} i_608
 * @param {JsonProducer} out_609
 * @returns {globalThis.number}
 */
function parseJsonBoolean_360(sourceText_607, i_608, out_609) {
  let return_610;
  let t_611;
  fn_612: {
    let ch0_613;
    if (sourceText_607.length > i_608) {
      t_611 = stringGet_356(sourceText_607, i_608);
      ch0_613 = t_611;
    } else {
      ch0_613 = 0;
    }
    const end_614 = sourceText_607.length;
    let keyword_615;
    let n_616;
    if (ch0_613 === 102) {
      keyword_615 = "false";
      n_616 = 5;
    } else if (ch0_613 === 116) {
      keyword_615 = "true";
      n_616 = 4;
    } else {
      keyword_615 = null;
      n_616 = 0;
    }
    if (!(keyword_615 == null)) {
      const keyword_617 = keyword_615;
      if (stringHasAtLeast_535(sourceText_607, i_608, end_614, n_616)) {
        const after_618 = afterSubstring_597(sourceText_607, i_608, keyword_617);
        if (after_618 >= 0) {
          return_610 = requireStringIndex_569(after_618);
          out_609.booleanValue(n_616 === 4);
          break fn_612;
        }
      }
    }
    expectedTokenError_355(sourceText_607, i_608, out_609, "`false` or `true`");
    return_610 = -1;
  }
  return return_610;
}
/**
 * @param {string} sourceText_619
 * @param {globalThis.number} i_620
 * @param {JsonProducer} out_621
 * @returns {globalThis.number}
 */
function parseJsonNull_361(sourceText_619, i_620, out_621) {
  let return_622;
  fn_623: {
    const after_624 = afterSubstring_597(sourceText_619, i_620, "null");
    if (after_624 >= 0) {
      return_622 = requireStringIndex_569(after_624);
      out_621.nullValue();
      break fn_623;
    }
    expectedTokenError_355(sourceText_619, i_620, out_621, "`null`");
    return_622 = -1;
  }
  return return_622;
}
/**
 * @param {string} sourceText_625
 * @param {globalThis.number} i_626
 * @param {JsonProducer} out_627
 * @returns {globalThis.number}
 */
function parseJsonNumber_362(sourceText_625, i_626, out_627) {
  let return_628;
  let t_629;
  let t_630;
  let t_631;
  let t_632;
  let t_633;
  let t_634;
  let t_635;
  let t_636;
  let t_637;
  let t_638;
  let t_639;
  let t_640;
  let t_641;
  let t_642;
  let t_643;
  let t_644;
  let t_645;
  let t_646;
  let t_647;
  let t_648;
  let t_649;
  let t_650;
  let t_651;
  let t_652;
  let t_653;
  let t_654;
  let t_655;
  let t_656;
  let t_657;
  let t_658;
  let t_659;
  let t_660;
  let t_661;
  let t_662;
  let t_663;
  let t_664;
  let t_665;
  let t_666;
  let t_667;
  fn_668: {
    let isNegative_669 = false;
    const startOfNumber_670 = i_626;
    if (sourceText_625.length > i_626) {
      t_629 = stringGet_356(sourceText_625, i_626);
      t_647 = t_629 === 45;
    } else {
      t_647 = false;
    }
    if (t_647) {
      isNegative_669 = true;
      t_630 = stringNext_464(sourceText_625, i_626);
      i_626 = t_630;
    }
    let digit0_671;
    if (sourceText_625.length > i_626) {
      t_631 = stringGet_356(sourceText_625, i_626);
      digit0_671 = t_631;
    } else {
      digit0_671 = -1;
    }
    if (digit0_671 < 48) {
      t_648 = true;
    } else {
      t_648 = 57 < digit0_671;
    }
    if (t_648) {
      let error_672;
      if (! isNegative_669) {
        t_649 = digit0_671 !== 46;
      } else {
        t_649 = false;
      }
      if (t_649) {
        error_672 = "JSON value";
      } else {
        error_672 = "digit";
      }
      expectedTokenError_355(sourceText_625, i_626, out_627, error_672);
      return_628 = -1;
      break fn_668;
    }
    t_632 = stringNext_464(sourceText_625, i_626);
    i_626 = t_632;
    let nDigits_673 = 1;
    t_633 = digit0_671 - 48 | 0;
    let tentativeFloat64_674 = t_633;
    t_634 = BigInt(digit0_671 - 48 | 0);
    let tentativeInt64_675 = t_634;
    let overflowInt64_676 = false;
    if (48 !== digit0_671) {
      while (true) {
        if (!(sourceText_625.length > i_626)) {
          break;
        }
        const possibleDigit_677 = stringGet_356(sourceText_625, i_626);
        if (48 <= possibleDigit_677) {
          t_650 = possibleDigit_677 <= 57;
        } else {
          t_650 = false;
        }
        if (t_650) {
          t_635 = stringNext_464(sourceText_625, i_626);
          i_626 = t_635;
          nDigits_673 = nDigits_673 + 1 | 0;
          const nextDigit_678 = possibleDigit_677 - 48 | 0;
          t_651 = tentativeFloat64_674 * 10.0;
          t_636 = nextDigit_678;
          tentativeFloat64_674 = t_651 + t_636;
          const oldInt64_679 = tentativeInt64_675;
          t_652 = clampInt64__680(tentativeInt64_675 * BigInt("10"));
          t_637 = BigInt(nextDigit_678);
          tentativeInt64_675 = clampInt64__680(t_652 + t_637);
          if (tentativeInt64_675 < oldInt64_679) {
            if (clampInt64__680(BigInt("-9223372036854775808") - oldInt64_679) === clampInt64__680(- BigInt(nextDigit_678))) {
              if (isNegative_669) {
                t_653 = oldInt64_679 > BigInt("0");
              } else {
                t_653 = false;
              }
              t_654 = t_653;
            } else {
              t_654 = false;
            }
            if (! t_654) {
              overflowInt64_676 = true;
            }
          }
        } else {
          break;
        }
      }
    }
    let nDigitsAfterPoint_681 = 0;
    if (sourceText_625.length > i_626) {
      t_638 = stringGet_356(sourceText_625, i_626);
      t_655 = 46 === t_638;
    } else {
      t_655 = false;
    }
    if (t_655) {
      t_639 = stringNext_464(sourceText_625, i_626);
      i_626 = t_639;
      const afterPoint_682 = i_626;
      while (true) {
        if (!(sourceText_625.length > i_626)) {
          break;
        }
        const possibleDigit_683 = stringGet_356(sourceText_625, i_626);
        if (48 <= possibleDigit_683) {
          t_656 = possibleDigit_683 <= 57;
        } else {
          t_656 = false;
        }
        if (t_656) {
          t_640 = stringNext_464(sourceText_625, i_626);
          i_626 = t_640;
          nDigits_673 = nDigits_673 + 1 | 0;
          nDigitsAfterPoint_681 = nDigitsAfterPoint_681 + 1 | 0;
          t_657 = tentativeFloat64_674 * 10.0;
          t_641 = possibleDigit_683 - 48 | 0;
          tentativeFloat64_674 = t_657 + t_641;
        } else {
          break;
        }
      }
      if (i_626 === afterPoint_682) {
        expectedTokenError_355(sourceText_625, i_626, out_627, "digit");
        return_628 = -1;
        break fn_668;
      }
    }
    let nExponentDigits_684 = 0;
    if (sourceText_625.length > i_626) {
      t_642 = stringGet_356(sourceText_625, i_626);
      t_658 = 101 ===(t_642 | 32);
    } else {
      t_658 = false;
    }
    if (t_658) {
      t_643 = stringNext_464(sourceText_625, i_626);
      i_626 = t_643;
      if (!(sourceText_625.length > i_626)) {
        expectedTokenError_355(sourceText_625, i_626, out_627, "sign or digit");
        return_628 = -1;
        break fn_668;
      }
      const afterE_685 = stringGet_356(sourceText_625, i_626);
      if (afterE_685 === 43) {
        t_659 = true;
      } else {
        t_659 = afterE_685 === 45;
      }
      if (t_659) {
        t_644 = stringNext_464(sourceText_625, i_626);
        i_626 = t_644;
      }
      while (true) {
        if (!(sourceText_625.length > i_626)) {
          break;
        }
        const possibleDigit_686 = stringGet_356(sourceText_625, i_626);
        if (48 <= possibleDigit_686) {
          t_660 = possibleDigit_686 <= 57;
        } else {
          t_660 = false;
        }
        if (t_660) {
          t_645 = stringNext_464(sourceText_625, i_626);
          i_626 = t_645;
          nExponentDigits_684 = nExponentDigits_684 + 1 | 0;
        } else {
          break;
        }
      }
      if (nExponentDigits_684 === 0) {
        expectedTokenError_355(sourceText_625, i_626, out_627, "exponent digit");
        return_628 = -1;
        break fn_668;
      }
    }
    const afterExponent_687 = i_626;
    if (nExponentDigits_684 === 0) {
      if (nDigitsAfterPoint_681 === 0) {
        t_661 = ! overflowInt64_676;
      } else {
        t_661 = false;
      }
      t_662 = t_661;
    } else {
      t_662 = false;
    }
    if (t_662) {
      let value_688;
      if (isNegative_669) {
        value_688 = clampInt64__680(- tentativeInt64_675);
      } else {
        value_688 = tentativeInt64_675;
      }
      if (BigInt("-2147483648") <= value_688) {
        t_663 = value_688 <= BigInt("2147483647");
      } else {
        t_663 = false;
      }
      if (t_663) {
        t_646 = int64ToInt32Unsafe_689(value_688);
        out_627.int32Value(t_646);
      } else {
        out_627.int64Value(value_688);
      }
      return_628 = i_626;
      break fn_668;
    }
    const numericTokenString_690 = sourceText_625.substring(startOfNumber_670, i_626);
    let doubleValue_691 = NaN;
    if (nExponentDigits_684 !== 0) {
      t_664 = true;
    } else {
      t_664 = nDigitsAfterPoint_681 !== 0;
    }
    if (t_664) {
      try {
        t_665 = stringToFloat64_192(numericTokenString_690);
        doubleValue_691 = t_665;
      } catch {
      }
    }
    if (cmpFloat__692(doubleValue_691, -Infinity) !== 0) {
      if (cmpFloat__692(doubleValue_691, Infinity) !== 0) {
        t_666 = cmpFloat__692(doubleValue_691, NaN) !== 0;
      } else {
        t_666 = false;
      }
      t_667 = t_666;
    } else {
      t_667 = false;
    }
    if (t_667) {
      out_627.float64Value(doubleValue_691);
    } else {
      out_627.numericTokenValue(numericTokenString_690);
    }
    return_628 = i_626;
  }
  return return_628;
}
/**
 * @param {string} sourceText_693
 * @param {JsonProducer} out_694
 */
export function parseJsonToProducer(sourceText_693, out_694) {
  let t_695;
  let t_696;
  let t_697;
  let t_698;
  let t_699;
  let t_700;
  let i_701 = 0;
  const afterValue_702 = parseJsonValue_345(sourceText_693, i_701, out_694);
  if (afterValue_702 >= 0) {
    t_700 = requireStringIndex_569(afterValue_702);
    t_695 = skipJsonSpaces_354(sourceText_693, t_700);
    i_701 = t_695;
    if (sourceText_693.length > i_701) {
      t_696 = out_694.parseErrorReceiver;
      t_699 = !(t_696 == null);
    } else {
      t_699 = false;
    }
    if (t_699) {
      t_697 = sourceText_693.length;
      t_698 = sourceText_693.substring(i_701, t_697);
      storeJsonError_465(out_694, "Extraneous JSON `" + t_698 + "`");
    }
  }
  return;
};
/**
 * @param {string} sourceText_703
 * @returns {JsonSyntaxTree}
 */
export function parseJson(sourceText_703) {
  const p_704 = new JsonSyntaxTreeProducer();
  parseJsonToProducer(sourceText_703, p_704);
  return p_704.toJsonSyntaxTree();
};
/** @returns {JsonAdapter<boolean>} */
export function booleanJsonAdapter() {
  return new BooleanJsonAdapter_370();
};
/** @returns {JsonAdapter<number>} */
export function float64JsonAdapter() {
  return new Float64JsonAdapter_378();
};
/** @returns {JsonAdapter<number>} */
export function int32JsonAdapter() {
  return new Int32JsonAdapter_386();
};
/** @returns {JsonAdapter<bigint>} */
export function int64JsonAdapter() {
  return new Int64JsonAdapter_394();
};
/** @returns {JsonAdapter<string>} */
export function stringJsonAdapter() {
  return new StringJsonAdapter_402();
};
/**
 * @template {unknown} T_706
 * @param {JsonAdapter<T_706>} adapterForT_705
 * @returns {JsonAdapter<Array<T_706>>}
 */
export function listJsonAdapter(adapterForT_705) {
  return new ListJsonAdapter_410(adapterForT_705);
};
