// @temperlang/std/regex.js
import {
  type as type__735, requireInstanceOf as requireInstanceOf__859, pairConstructor as pairConstructor_813, mapConstructor as mapConstructor_812, regexCompileFormatted as regexCompileFormatted_825, regexCompiledFound as regexCompiledFound_829, regexCompiledFind as regexCompiledFind_834, regexCompiledReplace as regexCompiledReplace_839, regexCompiledSplit as regexCompiledSplit_842, listedGet as listedGet_894, stringFromCodePoint as stringFromCodePoint_896, regexFormatterPushCodeTo as regexFormatterPushCodeTo_897, stringGet as stringGet_905, stringNext as stringNext_906, regexFormatterAdjustCodeSet as regexFormatterAdjustCodeSet_920, listBuilderAdd as listBuilderAdd_1029, listBuilderToList as listBuilderToList_1030
} from "@temperlang/core";
export class RegexNode extends type__735() {
  /** @returns {Regex} */
  compiled() {
    return new Regex(this);
  }
  /**
   * @param {string} text_727
   * @returns {boolean}
   */
  found(text_727) {
    return this.compiled().found(text_727);
  }
  /**
   * @param {string} text_729
   * @returns {Match}
   */
  find(text_729) {
    return this.compiled().find(text_729);
  }
  /**
   * @param {string} text_731
   * @param {(arg0: Match) => string} format_732
   * @returns {string}
   */
  replace(text_731, format_732) {
    return this.compiled().replace(text_731, format_732);
  }
  /**
   * @param {string} text_734
   * @returns {Array<string>}
   */
  split(text_734) {
    return this.compiled().split(text_734);
  }
};
export class Capture extends type__735(RegexNode) {
  /** @type {string} */
  #name_736;
  /** @type {RegexNode} */
  #item_737;
  /**
   * @param {{
   *   name: string, item: RegexNode
   * }}
   * props
   * @returns {Capture}
   */
  static["new"](props) {
    return new Capture(props.name, props.item);
  }
  /**
   * @param {string} name_738
   * @param {RegexNode} item_739
   */
  constructor(name_738, item_739) {
    super ();
    this.#name_736 = name_738;
    this.#item_737 = item_739;
    return;
  }
  /** @returns {string} */
  get name() {
    return this.#name_736;
  }
  /** @returns {RegexNode} */
  get item() {
    return this.#item_737;
  }
};
export class CodePart extends type__735(RegexNode) {
};
export class CodePoints extends type__735(CodePart) {
  /** @type {string} */
  #value_742;
  /** @param {string} value_743 */
  constructor(value_743) {
    super ();
    this.#value_742 = value_743;
    return;
  }
  /** @returns {string} */
  get value() {
    return this.#value_742;
  }
};
export class Special extends type__735(RegexNode) {
};
export class SpecialSet extends type__735(CodePart, Special) {
};
export class CodeRange extends type__735(CodePart) {
  /** @type {number} */
  #min_745;
  /** @type {number} */
  #max_746;
  /**
   * @param {{
   *   min: number, max: number
   * }}
   * props
   * @returns {CodeRange}
   */
  static["new"](props) {
    return new CodeRange(props.min, props.max);
  }
  /**
   * @param {number} min_747
   * @param {number} max_748
   */
  constructor(min_747, max_748) {
    super ();
    this.#min_745 = min_747;
    this.#max_746 = max_748;
    return;
  }
  /** @returns {number} */
  get min() {
    return this.#min_745;
  }
  /** @returns {number} */
  get max() {
    return this.#max_746;
  }
};
export class CodeSet extends type__735(RegexNode) {
  /** @type {Array<CodePart>} */
  #items_751;
  /** @type {boolean} */
  #negated_752;
  /**
   * @param {{
   *   items: Array<CodePart>, negated ?: boolean | null
   * }}
   * props
   * @returns {CodeSet}
   */
  static["new"](props) {
    return new CodeSet(props.items, props.negated);
  }
  /**
   * @param {Array<CodePart>} items_753
   * @param {boolean | null} [negated_754]
   */
  constructor(items_753, negated_754) {
    super ();
    let negated_755;
    if (negated_754 == null) {
      negated_755 = false;
    } else {
      negated_755 = negated_754;
    }
    this.#items_751 = items_753;
    this.#negated_752 = negated_755;
    return;
  }
  /** @returns {Array<CodePart>} */
  get items() {
    return this.#items_751;
  }
  /** @returns {boolean} */
  get negated() {
    return this.#negated_752;
  }
};
export class Or extends type__735(RegexNode) {
  /** @type {Array<RegexNode>} */
  #items_758;
  /** @param {Array<RegexNode>} items_759 */
  constructor(items_759) {
    super ();
    this.#items_758 = items_759;
    return;
  }
  /** @returns {Array<RegexNode>} */
  get items() {
    return this.#items_758;
  }
};
export class Repeat extends type__735(RegexNode) {
  /** @type {RegexNode} */
  #item_761;
  /** @type {number} */
  #min_762;
  /** @type {number | null} */
  #max_763;
  /** @type {boolean} */
  #reluctant_764;
  /**
   * @param {{
   *   item: RegexNode, min: number, max: number | null, reluctant ?: boolean | null
   * }}
   * props
   * @returns {Repeat}
   */
  static["new"](props) {
    return new Repeat(props.item, props.min, props.max, props.reluctant);
  }
  /**
   * @param {RegexNode} item_765
   * @param {number} min_766
   * @param {number | null} max_767
   * @param {boolean | null} [reluctant_768]
   */
  constructor(item_765, min_766, max_767, reluctant_768) {
    super ();
    let reluctant_769;
    if (reluctant_768 == null) {
      reluctant_769 = false;
    } else {
      reluctant_769 = reluctant_768;
    }
    this.#item_761 = item_765;
    this.#min_762 = min_766;
    this.#max_763 = max_767;
    this.#reluctant_764 = reluctant_769;
    return;
  }
  /** @returns {RegexNode} */
  get item() {
    return this.#item_761;
  }
  /** @returns {number} */
  get min() {
    return this.#min_762;
  }
  /** @returns {number | null} */
  get max() {
    return this.#max_763;
  }
  /** @returns {boolean} */
  get reluctant() {
    return this.#reluctant_764;
  }
};
export class Sequence extends type__735(RegexNode) {
  /** @type {Array<RegexNode>} */
  #items_774;
  /** @param {Array<RegexNode>} items_775 */
  constructor(items_775) {
    super ();
    this.#items_774 = items_775;
    return;
  }
  /** @returns {Array<RegexNode>} */
  get items() {
    return this.#items_774;
  }
};
export class Match extends type__735() {
  /** @type {Group} */
  #full_777;
  /** @type {Map<string, Group>} */
  #groups_778;
  /**
   * @param {{
   *   full: Group, groups: Map<string, Group>
   * }}
   * props
   * @returns {Match}
   */
  static["new"](props) {
    return new Match(props.full, props.groups);
  }
  /**
   * @param {Group} full_779
   * @param {Map<string, Group>} groups_780
   */
  constructor(full_779, groups_780) {
    super ();
    this.#full_777 = full_779;
    this.#groups_778 = groups_780;
    return;
  }
  /** @returns {Group} */
  get full() {
    return this.#full_777;
  }
  /** @returns {Map<string, Group>} */
  get groups() {
    return this.#groups_778;
  }
};
export class Group extends type__735() {
  /** @type {string} */
  #name_783;
  /** @type {string} */
  #value_784;
  /** @type {globalThis.number} */
  #begin_785;
  /** @type {globalThis.number} */
  #end_786;
  /**
   * @param {{
   *   name: string, value: string, begin: globalThis.number, end: globalThis.number
   * }}
   * props
   * @returns {Group}
   */
  static["new"](props) {
    return new Group(props.name, props.value, props.begin, props.end);
  }
  /**
   * @param {string} name_787
   * @param {string} value_788
   * @param {globalThis.number} begin_789
   * @param {globalThis.number} end_790
   */
  constructor(name_787, value_788, begin_789, end_790) {
    super ();
    this.#name_783 = name_787;
    this.#value_784 = value_788;
    this.#begin_785 = begin_789;
    this.#end_786 = end_790;
    return;
  }
  /** @returns {string} */
  get name() {
    return this.#name_783;
  }
  /** @returns {string} */
  get value() {
    return this.#value_784;
  }
  /** @returns {globalThis.number} */
  get begin() {
    return this.#begin_785;
  }
  /** @returns {globalThis.number} */
  get end() {
    return this.#end_786;
  }
};
class RegexRefs_795 extends type__735() {
  /** @type {CodePoints} */
  #codePoints_796;
  /** @type {Group} */
  #group_797;
  /** @type {Match} */
  #match_798;
  /** @type {Or} */
  #orObject_799;
  /**
   * @param {{
   *   codePoints ?: CodePoints | null, group ?: Group | null, match ?: Match | null, orObject ?: Or | null
   * }}
   * props
   * @returns {RegexRefs_795}
   */
  static["new"](props) {
    return new RegexRefs_795(props.codePoints, props.group, props.match, props.orObject);
  }
  /**
   * @param {CodePoints | null} [codePoints_800]
   * @param {Group | null} [group_801]
   * @param {Match | null} [match_802]
   * @param {Or | null} [orObject_803]
   */
  constructor(codePoints_800, group_801, match_802, orObject_803) {
    super ();
    let t_804;
    let t_805;
    let t_806;
    let t_807;
    let t_808;
    let codePoints_809;
    if (codePoints_800 == null) {
      t_804 = new CodePoints("");
      codePoints_809 = t_804;
    } else {
      codePoints_809 = codePoints_800;
    }
    let group_810;
    if (group_801 == null) {
      t_805 = new Group("", "", 0, 0);
      group_810 = t_805;
    } else {
      group_810 = group_801;
    }
    let match_811;
    if (match_802 == null) {
      t_806 = mapConstructor_812(Object.freeze([pairConstructor_813("", group_810)]));
      t_807 = new Match(group_810, t_806);
      match_811 = t_807;
    } else {
      match_811 = match_802;
    }
    let orObject_814;
    if (orObject_803 == null) {
      t_808 = new Or(Object.freeze([]));
      orObject_814 = t_808;
    } else {
      orObject_814 = orObject_803;
    }
    this.#codePoints_796 = codePoints_809;
    this.#group_797 = group_810;
    this.#match_798 = match_811;
    this.#orObject_799 = orObject_814;
    return;
  }
  /** @returns {CodePoints} */
  get codePoints() {
    return this.#codePoints_796;
  }
  /** @returns {Group} */
  get group() {
    return this.#group_797;
  }
  /** @returns {Match} */
  get match() {
    return this.#match_798;
  }
  /** @returns {Or} */
  get orObject() {
    return this.#orObject_799;
  }
}
export class Regex extends type__735() {
  /** @type {RegexNode} */
  #data_819;
  /** @param {RegexNode} data_820 */
  constructor(data_820) {
    super ();
    const t_821 = data_820;
    this.#data_819 = t_821;
    const formatted_822 = RegexFormatter_823.regexFormat(data_820);
    let t_824 = regexCompileFormatted_825(data_820, formatted_822);
    this.#compiled_826 = t_824;
    return;
  }
  /**
   * @param {string} text_828
   * @returns {boolean}
   */
  found(text_828) {
    return regexCompiledFound_829(this, this.#compiled_826, text_828);
  }
  /**
   * @param {string} text_831
   * @param {globalThis.number | null} [begin_832]
   * @returns {Match}
   */
  find(text_831, begin_832) {
    let begin_833;
    if (begin_832 == null) {
      begin_833 = 0;
    } else {
      begin_833 = begin_832;
    }
    return regexCompiledFind_834(this, this.#compiled_826, text_831, begin_833, regexRefs_835);
  }
  /**
   * @param {string} text_837
   * @param {(arg0: Match) => string} format_838
   * @returns {string}
   */
  replace(text_837, format_838) {
    return regexCompiledReplace_839(this, this.#compiled_826, text_837, format_838, regexRefs_835);
  }
  /**
   * @param {string} text_841
   * @returns {Array<string>}
   */
  split(text_841) {
    return regexCompiledSplit_842(this, this.#compiled_826, text_841, regexRefs_835);
  }
  /** @type {unknown} */
  #compiled_826;
  /** @returns {RegexNode} */
  get data() {
    return this.#data_819;
  }
};
class RegexFormatter_823 extends type__735() {
  /** @type {globalThis.Array<string>} */
  #out_844;
  /**
   * @param {RegexNode} data_846
   * @returns {string}
   */
  static regexFormat(data_846) {
    return new RegexFormatter_823().format(data_846);
  }
  /**
   * @param {RegexNode} regex_848
   * @returns {string}
   */
  format(regex_848) {
    this.#pushRegex_849(regex_848);
    return this.#out_844[0];
  }
  /** @param {RegexNode} regex_851 */
  #pushRegex_849(regex_851) {
    let t_852;
    let t_853;
    let t_854;
    let t_855;
    let t_856;
    let t_857;
    let t_858;
    if (regex_851 instanceof Capture) {
      t_852 = requireInstanceOf__859(regex_851, Capture);
      this.#pushCapture_860(t_852);
    } else if (regex_851 instanceof CodePoints) {
      t_853 = requireInstanceOf__859(regex_851, CodePoints);
      this.#pushCodePoints_861(t_853, false);
    } else if (regex_851 instanceof CodeRange) {
      t_854 = requireInstanceOf__859(regex_851, CodeRange);
      this.#pushCodeRange_862(t_854);
    } else if (regex_851 instanceof CodeSet) {
      t_855 = requireInstanceOf__859(regex_851, CodeSet);
      this.#pushCodeSet_863(t_855);
    } else if (regex_851 instanceof Or) {
      t_856 = requireInstanceOf__859(regex_851, Or);
      this.#pushOr_864(t_856);
    } else if (regex_851 instanceof Repeat) {
      t_857 = requireInstanceOf__859(regex_851, Repeat);
      this.#pushRepeat_865(t_857);
    } else if (regex_851 instanceof Sequence) {
      t_858 = requireInstanceOf__859(regex_851, Sequence);
      this.#pushSequence_866(t_858);
    } else if (Object.is(regex_851, Begin)) {
      this.#out_844[0] += "^";
    } else if (Object.is(regex_851, Dot)) {
      this.#out_844[0] += ".";
    } else if (Object.is(regex_851, End)) {
      this.#out_844[0] += "\u0024";
    } else if (Object.is(regex_851, WordBoundary)) {
      this.#out_844[0] += "\\b";
    } else if (Object.is(regex_851, Digit)) {
      this.#out_844[0] += "\\d";
    } else if (Object.is(regex_851, Space)) {
      this.#out_844[0] += "\\s";
    } else if (Object.is(regex_851, Word)) {
      this.#out_844[0] += "\\w";
    }
    return;
  }
  /** @param {Capture} capture_868 */
  #pushCapture_860(capture_868) {
    this.#out_844[0] += "(";
    let t_869 = this.#out_844;
    let t_870 = capture_868.name;
    this.#pushCaptureName_871(t_869, t_870);
    let t_872 = capture_868.item;
    this.#pushRegex_849(t_872);
    this.#out_844[0] += ")";
    return;
  }
  /**
   * @param {globalThis.Array<string>} out_874
   * @param {string} name_875
   */
  #pushCaptureName_871(out_874, name_875) {
    out_874[0] += "?\u003c" + name_875 + "\u003e";
    return;
  }
  /**
   * @param {number} code_878
   * @param {boolean} insideCodeSet_879
   */
  #pushCode_877(code_878, insideCodeSet_879) {
    let return_880;
    let t_881;
    let t_882;
    let t_883;
    let t_884;
    let t_885;
    let t_886;
    let t_887;
    let t_888;
    let t_889;
    fn_890: {
      try {
        let specialEscape_891;
        if (code_878 === Codes_892.carriageReturn) {
          specialEscape_891 = "r";
        } else if (code_878 === Codes_892.newline) {
          specialEscape_891 = "n";
        } else if (code_878 === Codes_892.tab) {
          specialEscape_891 = "t";
        } else {
          specialEscape_891 = "";
        }
        if (specialEscape_891 !== "") {
          this.#out_844[0] += "\\";
          this.#out_844[0] += specialEscape_891;
          return_880 = void 0;
          break fn_890;
        }
        if (code_878 <= 127) {
          const escapeNeed_893 = listedGet_894(escapeNeeds_895, code_878);
          if (escapeNeed_893 === 2) {
            t_882 = true;
          } else {
            if (insideCodeSet_879) {
              t_881 = code_878 === Codes_892.dash;
            } else {
              t_881 = false;
            }
            t_882 = t_881;
          }
          if (t_882) {
            this.#out_844[0] += "\\";
            t_883 = stringFromCodePoint_896(code_878);
            this.#out_844[0] += t_883;
            return_880 = void 0;
            break fn_890;
          } else if (escapeNeed_893 === 0) {
            t_884 = stringFromCodePoint_896(code_878);
            this.#out_844[0] += t_884;
            return_880 = void 0;
            break fn_890;
          }
        }
        if (code_878 >= Codes_892.supplementalMin) {
          t_888 = true;
        } else {
          if (code_878 > Codes_892.highControlMax) {
            if (Codes_892.surrogateMin <= code_878) {
              t_885 = code_878 <= Codes_892.surrogateMax;
            } else {
              t_885 = false;
            }
            if (t_885) {
              t_886 = true;
            } else {
              t_886 = code_878 === Codes_892.uint16Max;
            }
            t_887 = ! t_886;
          } else {
            t_887 = false;
          }
          t_888 = t_887;
        }
        if (t_888) {
          t_889 = stringFromCodePoint_896(code_878);
          this.#out_844[0] += t_889;
        } else {
          regexFormatterPushCodeTo_897(this, this.#out_844, code_878, insideCodeSet_879);
        }
      } catch {
        throw Error();
      }
      return_880 = void 0;
    }
    return return_880;
  }
  /**
   * @param {CodePoints} codePoints_899
   * @param {boolean} insideCodeSet_900
   */
  #pushCodePoints_861(codePoints_899, insideCodeSet_900) {
    let t_901;
    let t_902;
    const value_903 = codePoints_899.value;
    let index_904 = 0;
    while (true) {
      if (!(value_903.length > index_904)) {
        break;
      }
      t_901 = stringGet_905(value_903, index_904);
      this.#pushCode_877(t_901, insideCodeSet_900);
      t_902 = stringNext_906(value_903, index_904);
      index_904 = t_902;
    }
    return;
  }
  /** @param {CodeRange} codeRange_908 */
  #pushCodeRange_862(codeRange_908) {
    this.#out_844[0] += "[";
    this.#pushCodeRangeUnwrapped_909(codeRange_908);
    this.#out_844[0] += "]";
    return;
  }
  /** @param {CodeRange} codeRange_911 */
  #pushCodeRangeUnwrapped_909(codeRange_911) {
    let t_912 = codeRange_911.min;
    this.#pushCode_877(t_912, true);
    this.#out_844[0] += "-";
    let t_913 = codeRange_911.max;
    this.#pushCode_877(t_913, true);
    return;
  }
  /** @param {CodeSet} codeSet_915 */
  #pushCodeSet_863(codeSet_915) {
    let t_916;
    let t_917;
    let t_918;
    const adjusted_919 = regexFormatterAdjustCodeSet_920(this, codeSet_915, regexRefs_835);
    if (adjusted_919 instanceof CodeSet) {
      t_918 = requireInstanceOf__859(adjusted_919, CodeSet);
      this.#out_844[0] += "[";
      if (t_918.negated) {
        this.#out_844[0] += "^";
      }
      let i_921 = 0;
      while (true) {
        t_916 = t_918.items.length;
        if (!(i_921 < t_916)) {
          break;
        }
        t_917 = listedGet_894(t_918.items, i_921);
        this.#pushCodeSetItem_922(t_917);
        i_921 = i_921 + 1 | 0;
      }
      this.#out_844[0] += "]";
    } else {
      this.#pushRegex_849(adjusted_919);
    }
    return;
  }
  /** @param {CodePart} codePart_924 */
  #pushCodeSetItem_922(codePart_924) {
    let t_925;
    let t_926;
    let t_927;
    if (codePart_924 instanceof CodePoints) {
      t_925 = requireInstanceOf__859(codePart_924, CodePoints);
      this.#pushCodePoints_861(t_925, true);
    } else if (codePart_924 instanceof CodeRange) {
      t_926 = requireInstanceOf__859(codePart_924, CodeRange);
      this.#pushCodeRangeUnwrapped_909(t_926);
    } else if (codePart_924 instanceof SpecialSet) {
      t_927 = requireInstanceOf__859(codePart_924, SpecialSet);
      this.#pushRegex_849(t_927);
    }
    return;
  }
  /** @param {Or} or_929 */
  #pushOr_864(or_929) {
    let t_930;
    let t_931;
    let t_932;
    if (! ! or_929.items.length) {
      this.#out_844[0] += "(?:";
      t_930 = listedGet_894(or_929.items, 0);
      this.#pushRegex_849(t_930);
      let i_933 = 1;
      while (true) {
        t_931 = or_929.items.length;
        if (!(i_933 < t_931)) {
          break;
        }
        this.#out_844[0] += "|";
        t_932 = listedGet_894(or_929.items, i_933);
        this.#pushRegex_849(t_932);
        i_933 = i_933 + 1 | 0;
      }
      this.#out_844[0] += ")";
    }
    return;
  }
  /** @param {Repeat} repeat_935 */
  #pushRepeat_865(repeat_935) {
    let t_936;
    let t_937;
    let t_938;
    let t_939;
    let t_940;
    this.#out_844[0] += "(?:";
    let t_941 = repeat_935.item;
    this.#pushRegex_849(t_941);
    this.#out_844[0] += ")";
    const min_942 = repeat_935.min;
    const max_943 = repeat_935.max;
    if (min_942 === 0) {
      t_938 = max_943 === 1;
    } else {
      t_938 = false;
    }
    if (t_938) {
      this.#out_844[0] += "?";
    } else {
      if (min_942 === 0) {
        t_939 = max_943 == null;
      } else {
        t_939 = false;
      }
      if (t_939) {
        this.#out_844[0] += "*";
      } else {
        if (min_942 === 1) {
          t_940 = max_943 == null;
        } else {
          t_940 = false;
        }
        if (t_940) {
          this.#out_844[0] += "+";
        } else {
          t_936 = min_942.toString();
          this.#out_844[0] += "{" + t_936;
          if (min_942 !== max_943) {
            this.#out_844[0] += ",";
            if (!(max_943 == null)) {
              t_937 = max_943.toString();
              this.#out_844[0] += t_937;
            }
          }
          this.#out_844[0] += "}";
        }
      }
    }
    if (repeat_935.reluctant) {
      this.#out_844[0] += "?";
    }
    return;
  }
  /** @param {Sequence} sequence_945 */
  #pushSequence_866(sequence_945) {
    let t_946;
    let t_947;
    let i_948 = 0;
    while (true) {
      t_946 = sequence_945.items.length;
      if (!(i_948 < t_946)) {
        break;
      }
      t_947 = listedGet_894(sequence_945.items, i_948);
      this.#pushRegex_849(t_947);
      i_948 = i_948 + 1 | 0;
    }
    return;
  }
  /**
   * @param {CodePart} codePart_950
   * @returns {number | null}
   */
  maxCode(codePart_950) {
    let return_951;
    let t_952;
    let t_953;
    if (codePart_950 instanceof CodePoints) {
      t_953 = requireInstanceOf__859(codePart_950, CodePoints);
      const value_954 = t_953.value;
      if (! value_954) {
        return_951 = null;
      } else {
        let max_955 = 0;
        let index_956 = 0;
        while (true) {
          if (!(value_954.length > index_956)) {
            break;
          }
          const next_957 = stringGet_905(value_954, index_956);
          if (next_957 > max_955) {
            max_955 = next_957;
          }
          t_952 = stringNext_906(value_954, index_956);
          index_956 = t_952;
        }
        return_951 = max_955;
      }
    } else if (codePart_950 instanceof CodeRange) {
      return_951 = requireInstanceOf__859(codePart_950, CodeRange).max;
    } else if (Object.is(codePart_950, Digit)) {
      return_951 = Codes_892.digit9;
    } else if (Object.is(codePart_950, Space)) {
      return_951 = Codes_892.space;
    } else if (Object.is(codePart_950, Word)) {
      return_951 = Codes_892.lowerZ;
    } else {
      return_951 = null;
    }
    return return_951;
  }
  constructor() {
    super ();
    let t_958 = [""];
    this.#out_844 = t_958;
    return;
  }
}
class Codes_892 extends type__735() {
  /** @type {number} */
  static #ampersand_959 = 38;
  /** @returns {number} */
  static get ampersand() {
    return this.#ampersand_959;
  }
  /** @type {number} */
  static #backslash_960 = 92;
  /** @returns {number} */
  static get backslash() {
    return this.#backslash_960;
  }
  /** @type {number} */
  static #caret_961 = 94;
  /** @returns {number} */
  static get caret() {
    return this.#caret_961;
  }
  /** @type {number} */
  static #carriageReturn_962 = 13;
  /** @returns {number} */
  static get carriageReturn() {
    return this.#carriageReturn_962;
  }
  /** @type {number} */
  static #curlyLeft_963 = 123;
  /** @returns {number} */
  static get curlyLeft() {
    return this.#curlyLeft_963;
  }
  /** @type {number} */
  static #curlyRight_964 = 125;
  /** @returns {number} */
  static get curlyRight() {
    return this.#curlyRight_964;
  }
  /** @type {number} */
  static #dash_965 = 45;
  /** @returns {number} */
  static get dash() {
    return this.#dash_965;
  }
  /** @type {number} */
  static #dot_966 = 46;
  /** @returns {number} */
  static get dot() {
    return this.#dot_966;
  }
  /** @type {number} */
  static #highControlMin_967 = 127;
  /** @returns {number} */
  static get highControlMin() {
    return this.#highControlMin_967;
  }
  /** @type {number} */
  static #highControlMax_968 = 159;
  /** @returns {number} */
  static get highControlMax() {
    return this.#highControlMax_968;
  }
  /** @type {number} */
  static #digit0_969 = 48;
  /** @returns {number} */
  static get digit0() {
    return this.#digit0_969;
  }
  /** @type {number} */
  static #digit9_970 = 57;
  /** @returns {number} */
  static get digit9() {
    return this.#digit9_970;
  }
  /** @type {number} */
  static #lowerA_971 = 97;
  /** @returns {number} */
  static get lowerA() {
    return this.#lowerA_971;
  }
  /** @type {number} */
  static #lowerZ_972 = 122;
  /** @returns {number} */
  static get lowerZ() {
    return this.#lowerZ_972;
  }
  /** @type {number} */
  static #newline_973 = 10;
  /** @returns {number} */
  static get newline() {
    return this.#newline_973;
  }
  /** @type {number} */
  static #peso_974 = 36;
  /** @returns {number} */
  static get peso() {
    return this.#peso_974;
  }
  /** @type {number} */
  static #pipe_975 = 124;
  /** @returns {number} */
  static get pipe() {
    return this.#pipe_975;
  }
  /** @type {number} */
  static #plus_976 = 43;
  /** @returns {number} */
  static get plus() {
    return this.#plus_976;
  }
  /** @type {number} */
  static #question_977 = 63;
  /** @returns {number} */
  static get question() {
    return this.#question_977;
  }
  /** @type {number} */
  static #roundLeft_978 = 40;
  /** @returns {number} */
  static get roundLeft() {
    return this.#roundLeft_978;
  }
  /** @type {number} */
  static #roundRight_979 = 41;
  /** @returns {number} */
  static get roundRight() {
    return this.#roundRight_979;
  }
  /** @type {number} */
  static #slash_980 = 47;
  /** @returns {number} */
  static get slash() {
    return this.#slash_980;
  }
  /** @type {number} */
  static #squareLeft_981 = 91;
  /** @returns {number} */
  static get squareLeft() {
    return this.#squareLeft_981;
  }
  /** @type {number} */
  static #squareRight_982 = 93;
  /** @returns {number} */
  static get squareRight() {
    return this.#squareRight_982;
  }
  /** @type {number} */
  static #star_983 = 42;
  /** @returns {number} */
  static get star() {
    return this.#star_983;
  }
  /** @type {number} */
  static #tab_984 = 9;
  /** @returns {number} */
  static get tab() {
    return this.#tab_984;
  }
  /** @type {number} */
  static #tilde_985 = 42;
  /** @returns {number} */
  static get tilde() {
    return this.#tilde_985;
  }
  /** @type {number} */
  static #upperA_986 = 65;
  /** @returns {number} */
  static get upperA() {
    return this.#upperA_986;
  }
  /** @type {number} */
  static #upperZ_987 = 90;
  /** @returns {number} */
  static get upperZ() {
    return this.#upperZ_987;
  }
  /** @type {number} */
  static #space_988 = 32;
  /** @returns {number} */
  static get space() {
    return this.#space_988;
  }
  /** @type {number} */
  static #surrogateMin_989 = 55296;
  /** @returns {number} */
  static get surrogateMin() {
    return this.#surrogateMin_989;
  }
  /** @type {number} */
  static #surrogateMax_990 = 57343;
  /** @returns {number} */
  static get surrogateMax() {
    return this.#surrogateMax_990;
  }
  /** @type {number} */
  static #supplementalMin_991 = 65536;
  /** @returns {number} */
  static get supplementalMin() {
    return this.#supplementalMin_991;
  }
  /** @type {number} */
  static #uint16Max_992 = 65535;
  /** @returns {number} */
  static get uint16Max() {
    return this.#uint16Max_992;
  }
  /** @type {number} */
  static #underscore_993 = 95;
  /** @returns {number} */
  static get underscore() {
    return this.#underscore_993;
  }
  constructor() {
    super ();
    return;
  }
}
class Begin_994 extends type__735(Special) {
  constructor() {
    super ();
    return;
  }
}
/** @type {Special} */
export const Begin = new Begin_994();
class Dot_995 extends type__735(Special) {
  constructor() {
    super ();
    return;
  }
}
/** @type {Special} */
export const Dot = new Dot_995();
class End_996 extends type__735(Special) {
  constructor() {
    super ();
    return;
  }
}
/** @type {Special} */
export const End = new End_996();
class WordBoundary_997 extends type__735(Special) {
  constructor() {
    super ();
    return;
  }
}
/** @type {Special} */
export const WordBoundary = new WordBoundary_997();
class Digit_998 extends type__735(SpecialSet) {
  constructor() {
    super ();
    return;
  }
}
/** @type {SpecialSet} */
export const Digit = new Digit_998();
class Space_999 extends type__735(SpecialSet) {
  constructor() {
    super ();
    return;
  }
}
/** @type {SpecialSet} */
export const Space = new Space_999();
class Word_1000 extends type__735(SpecialSet) {
  constructor() {
    super ();
    return;
  }
}
/** @type {SpecialSet} */
export const Word = new Word_1000();
/** @returns {Array<number>} */
function buildEscapeNeeds_1001() {
  let t_1002;
  let t_1003;
  let t_1004;
  let t_1005;
  let t_1006;
  let t_1007;
  let t_1008;
  let t_1009;
  let t_1010;
  let t_1011;
  let t_1012;
  let t_1013;
  let t_1014;
  let t_1015;
  let t_1016;
  let t_1017;
  let t_1018;
  let t_1019;
  let t_1020;
  let t_1021;
  let t_1022;
  let t_1023;
  let t_1024;
  let t_1025;
  let t_1026;
  const escapeNeeds_1027 = [];
  let code_1028 = 0;
  while (code_1028 < 127) {
    if (code_1028 === Codes_892.dash) {
      t_1009 = true;
    } else {
      if (code_1028 === Codes_892.space) {
        t_1008 = true;
      } else {
        if (code_1028 === Codes_892.underscore) {
          t_1007 = true;
        } else {
          if (Codes_892.digit0 <= code_1028) {
            t_1002 = code_1028 <= Codes_892.digit9;
          } else {
            t_1002 = false;
          }
          if (t_1002) {
            t_1006 = true;
          } else {
            if (Codes_892.upperA <= code_1028) {
              t_1003 = code_1028 <= Codes_892.upperZ;
            } else {
              t_1003 = false;
            }
            if (t_1003) {
              t_1005 = true;
            } else {
              if (Codes_892.lowerA <= code_1028) {
                t_1004 = code_1028 <= Codes_892.lowerZ;
              } else {
                t_1004 = false;
              }
              t_1005 = t_1004;
            }
            t_1006 = t_1005;
          }
          t_1007 = t_1006;
        }
        t_1008 = t_1007;
      }
      t_1009 = t_1008;
    }
    if (t_1009) {
      t_1026 = 0;
    } else {
      if (code_1028 === Codes_892.ampersand) {
        t_1025 = true;
      } else {
        if (code_1028 === Codes_892.backslash) {
          t_1024 = true;
        } else {
          if (code_1028 === Codes_892.caret) {
            t_1023 = true;
          } else {
            if (code_1028 === Codes_892.curlyLeft) {
              t_1022 = true;
            } else {
              if (code_1028 === Codes_892.curlyRight) {
                t_1021 = true;
              } else {
                if (code_1028 === Codes_892.dot) {
                  t_1020 = true;
                } else {
                  if (code_1028 === Codes_892.peso) {
                    t_1019 = true;
                  } else {
                    if (code_1028 === Codes_892.pipe) {
                      t_1018 = true;
                    } else {
                      if (code_1028 === Codes_892.plus) {
                        t_1017 = true;
                      } else {
                        if (code_1028 === Codes_892.question) {
                          t_1016 = true;
                        } else {
                          if (code_1028 === Codes_892.roundLeft) {
                            t_1015 = true;
                          } else {
                            if (code_1028 === Codes_892.roundRight) {
                              t_1014 = true;
                            } else {
                              if (code_1028 === Codes_892.slash) {
                                t_1013 = true;
                              } else {
                                if (code_1028 === Codes_892.squareLeft) {
                                  t_1012 = true;
                                } else {
                                  if (code_1028 === Codes_892.squareRight) {
                                    t_1011 = true;
                                  } else {
                                    if (code_1028 === Codes_892.star) {
                                      t_1010 = true;
                                    } else {
                                      t_1010 = code_1028 === Codes_892.tilde;
                                    }
                                    t_1011 = t_1010;
                                  }
                                  t_1012 = t_1011;
                                }
                                t_1013 = t_1012;
                              }
                              t_1014 = t_1013;
                            }
                            t_1015 = t_1014;
                          }
                          t_1016 = t_1015;
                        }
                        t_1017 = t_1016;
                      }
                      t_1018 = t_1017;
                    }
                    t_1019 = t_1018;
                  }
                  t_1020 = t_1019;
                }
                t_1021 = t_1020;
              }
              t_1022 = t_1021;
            }
            t_1023 = t_1022;
          }
          t_1024 = t_1023;
        }
        t_1025 = t_1024;
      }
      if (t_1025) {
        t_1026 = 2;
      } else {
        t_1026 = 1;
      }
    }
    listBuilderAdd_1029(escapeNeeds_1027, t_1026);
    code_1028 = code_1028 + 1 | 0;
  }
  return listBuilderToList_1030(escapeNeeds_1027);
}
/** @type {Array<number>} */
const escapeNeeds_895 = buildEscapeNeeds_1001();
/** @type {RegexRefs_795} */
const regexRefs_835 = new RegexRefs_795();
/**
 * @param {RegexNode} item_1031
 * @returns {RegexNode}
 */
export function entire(item_1031) {
  return new Sequence(Object.freeze([Begin, item_1031, End]));
};
/**
 * @param {RegexNode} item_1032
 * @param {boolean | null} [reluctant_1033]
 * @returns {Repeat}
 */
export function oneOrMore(item_1032, reluctant_1033) {
  let reluctant_1034;
  if (reluctant_1033 == null) {
    reluctant_1034 = false;
  } else {
    reluctant_1034 = reluctant_1033;
  }
  return new Repeat(item_1032, 1, null, reluctant_1034);
};
/**
 * @param {RegexNode} item_1035
 * @param {boolean | null} [reluctant_1036]
 * @returns {Repeat}
 */
export function optional(item_1035, reluctant_1036) {
  let reluctant_1037;
  if (reluctant_1036 == null) {
    reluctant_1037 = false;
  } else {
    reluctant_1037 = reluctant_1036;
  }
  return new Repeat(item_1035, 0, 1, reluctant_1037);
};
