// @temperlang/std/testing.js
import {
  strict as strict__11
} from "assert";
import {
  type as type__25, listBuilderAdd as listBuilderAdd_5, listBuilderToList as listBuilderToList_15, listedJoin as listedJoin_23, pairConstructor as pairConstructor_40, listedMap as listedMap_41, listedReduceFrom as listedReduceFrom_52, stringSplit as stringSplit_57, listedGet as listedGet_62
} from "@temperlang/core";
export class Test extends type__25() {
  /**
   * @param {boolean} success_1
   * @param {() => string} message_2
   */
  assert(success_1, message_2) {
    let t_3;
    if (! success_1) {
      this.#_passing_4 = false;
      t_3 = message_2();
      listBuilderAdd_5(this.#_messages_6, t_3);
    }
    return;
  }
  /**
   * @param {boolean} success_8
   * @param {() => string} message_9
   * @returns {void}
   */
  assertHard(success_8, message_9) {
    this.assert(success_8, message_9);
    if (! success_8) {
      this.#_failedOnAssert_10 = true;
      strict__11.fail(this.messagesCombined());
    }
    return;
  }
  /** @returns {void} */
  softFailToHard() {
    if (this.hasUnhandledFail) {
      this.#_failedOnAssert_10 = true;
      strict__11.fail(this.messagesCombined());
    }
    return;
  }
  /** @returns {boolean} */
  get passing() {
    return this.#_passing_4;
  }
  /** @returns {Array<string>} */
  messages() {
    return listBuilderToList_15(this.#_messages_6);
  }
  /** @returns {boolean} */
  get failedOnAssert() {
    return this.#_failedOnAssert_10;
  }
  /** @returns {boolean} */
  get hasUnhandledFail() {
    let t_18;
    if (this.#_failedOnAssert_10) {
      t_18 = true;
    } else {
      t_18 = this.#_passing_4;
    }
    return ! t_18;
  }
  /** @returns {string | null} */
  messagesCombined() {
    let return_20;
    if (! this.#_messages_6.length) {
      return_20 = null;
    } else {
      function fn_21(it_22) {
        return it_22;
      }
      return_20 = listedJoin_23(this.#_messages_6, ", ", fn_21);
    }
    return return_20;
  }
  /** @type {boolean} */
  #_failedOnAssert_10;
  /** @type {boolean} */
  #_passing_4;
  /** @type {Array<string>} */
  #_messages_6;
  constructor() {
    super ();
    this.#_failedOnAssert_10 = false;
    this.#_passing_4 = true;
    let t_24 = [];
    this.#_messages_6 = t_24;
    return;
  }
};
/**
 * @param {Array<Pair_42<string, (arg0: Test) => void>>} testCases_26
 * @returns {Array<Pair_42<string, Array<string>>>}
 */
export function processTestCases(testCases_26) {
  function fn_27(testCase_28) {
    let t_29;
    let t_30;
    let t_31;
    let t_32;
    const key_33 = testCase_28.key;
    const fun_34 = testCase_28.value;
    const test_35 = new Test();
    let hadBubble_36 = false;
    try {
      fun_34(test_35);
    } catch {
      hadBubble_36 = true;
    }
    const messages_37 = test_35.messages();
    let failures_38;
    if (test_35.passing) {
      t_31 = ! hadBubble_36;
    } else {
      t_31 = false;
    }
    if (t_31) {
      failures_38 = Object.freeze([]);
    } else {
      if (hadBubble_36) {
        t_29 = test_35.failedOnAssert;
        t_32 = ! t_29;
      } else {
        t_32 = false;
      }
      if (t_32) {
        const allMessages_39 = messages_37.slice();
        listBuilderAdd_5(allMessages_39, "Bubble");
        t_30 = listBuilderToList_15(allMessages_39);
        failures_38 = t_30;
      } else {
        failures_38 = messages_37;
      }
    }
    return pairConstructor_40(key_33, failures_38);
  }
  return listedMap_41(testCases_26, fn_27);
};
/**
 * @param {Array<Pair_42<string, Array<string>>>} testResults_43
 * @param {(arg0: string) => void} writeLine_44
 */
export function reportTestResults(testResults_43, writeLine_44) {
  let t_45;
  writeLine_44("\u003ctestsuites\u003e");
  const total_46 = testResults_43.length.toString();
  function fn_47(fails_48, testResult_49) {
    let t_50;
    if (! testResult_49.value.length) {
      t_50 = 0;
    } else {
      t_50 = 1;
    }
    return fails_48 + t_50 | 0;
  }
  const fails_51 = listedReduceFrom_52(testResults_43, 0, fn_47).toString();
  const totals_53 = "tests='" + total_46 + "' failures='" + fails_51 + "'";
  writeLine_44("  \u003ctestsuite name='suite' " + totals_53 + " time='0.0'\u003e");
  function escape_54(s_55) {
    let t_56 = stringSplit_57(s_55, "'");
    function fn_58(x_59) {
      return x_59;
    }
    return listedJoin_23(t_56, "\u0026apos;", fn_58);
  }
  let i_60 = 0;
  while (true) {
    t_45 = testResults_43.length;
    if (!(i_60 < t_45)) {
      break;
    }
    const testResult_61 = listedGet_62(testResults_43, i_60);
    const failureMessages_63 = testResult_61.value;
    const name_64 = escape_54(testResult_61.key);
    const basics_65 = "name='" + name_64 + "' classname='" + name_64 + "' time='0.0'";
    if (! failureMessages_63.length) {
      writeLine_44("    \u003ctestcase " + basics_65 + " /\u003e");
    } else {
      writeLine_44("    \u003ctestcase " + basics_65 + "\u003e");
      function fn_66(it_67) {
        return it_67;
      }
      const message_68 = escape_54(listedJoin_23(failureMessages_63, ", ", fn_66));
      writeLine_44("      \u003cfailure message='" + message_68 + "' /\u003e");
      writeLine_44("    \u003c/testcase\u003e");
    }
    i_60 = i_60 + 1 | 0;
  }
  writeLine_44("  \u003c/testsuite\u003e");
  writeLine_44("\u003c/testsuites\u003e");
  return;
};
/**
 * @param {Array<Pair_42<string, (arg0: Test) => void>>} testCases_69
 * @returns {string}
 */
export function runTestCases(testCases_69) {
  const report_70 = [""];
  let t_71 = processTestCases(testCases_69);
  function fn_72(line_73) {
    report_70[0] += line_73;
    report_70[0] += "\n";
    return;
  }
  reportTestResults(t_71, fn_72);
  return report_70[0];
};
/**
 * @param {(arg0: Test) => void} testFun_74
 * @returns {void}
 */
export function runTest(testFun_74) {
  const test_75 = new Test();
  try {
    testFun_74(test_75);
  } catch {
    function fn_76() {
      return "bubble during test running";
    }
    test_75.assert(false, fn_76);
  }
  test_75.softFailToHard();
  return;
};
